<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="DongyueXie" />

<meta name="date" content="2020-03-18" />

<title>Binomial Thinning Ultimate</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.0.13/css/fa-svg-with-js.css" rel="stylesheet" />
<script src="site_libs/font-awesome-5.0.13/js/fontawesome-all.min.js"></script>
<script src="site_libs/font-awesome-5.0.13/js/fa-v4-shims.min.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">misc</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/DongyueXie/misc">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Binomial Thinning Ultimate</h1>
<h4 class="author"><em>DongyueXie</em></h4>
<h4 class="date"><em>2020-03-18</em></h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-04-01
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>misc/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20191122code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20191122)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20191122code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20191122)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomDongyueXiemisctree42ee8caf9e46d7d04ffaa47fee7b1c3ef230cdc2targetblank42ee8caa"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/DongyueXie/misc/tree/42ee8caf9e46d7d04ffaa47fee7b1c3ef230cdc2" target="_blank">42ee8ca</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomDongyueXiemisctree42ee8caf9e46d7d04ffaa47fee7b1c3ef230cdc2targetblank42ee8caa" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  analysis/contrainedclustering.Rmd
    Untracked:  analysis/gsea.Rmd
    Untracked:  analysis/ideas.Rmd
    Untracked:  analysis/methylation.Rmd
    Untracked:  analysis/susie.Rmd
    Untracked:  code/sccytokines.R
    Untracked:  code/scdeCalibration.R
    Untracked:  data/bart/
    Untracked:  data/cytokine/DE_controls_output_filter10.RData
    Untracked:  data/cytokine/DE_controls_output_filter10_addlimma.RData
    Untracked:  data/cytokine/README
    Untracked:  data/cytokine/test.RData
    Untracked:  data/cytokine_normalized.RData
    Untracked:  data/deconv/
    Untracked:  data/scde/

Unstaged changes:
    Modified:   analysis/CPM.Rmd
    Modified:   analysis/deconvolution.Rmd
    Modified:   analysis/index.Rmd
    Modified:   analysis/limma.Rmd
    Deleted:    data/mout_high.RData
    Deleted:    data/scCDT.RData
    Deleted:    data/sva_sva_high.RData

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DongyueXie/misc/blob/42ee8caf9e46d7d04ffaa47fee7b1c3ef230cdc2/analysis/binomthinultimate.Rmd" target="_blank">42ee8ca</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-01
</td>
<td>
wflow_publish(“analysis/binomthinultimate.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DongyueXie/misc/7177fb223462fb9ca44ae519f0efc54467adea8c/docs/binomthinultimate.html" target="_blank">7177fb2</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-01
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DongyueXie/misc/blob/bfa42666f3bc8975d9b7f6649906861d51575408/analysis/binomthinultimate.Rmd" target="_blank">bfa4266</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-01
</td>
<td>
wflow_publish(“analysis/binomthinultimate.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DongyueXie/misc/9576644230dc49d2b963c24ef73b6802de727f63/docs/binomthinultimate.html" target="_blank">9576644</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-03-21
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DongyueXie/misc/blob/d6cf110432b833d65fa9609206848bea97d17b0f/analysis/binomthinultimate.Rmd" target="_blank">d6cf110</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-03-21
</td>
<td>
wflow_publish(“analysis/binomthinultimate.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DongyueXie/misc/9d09fc0cf50f1fcf370df115cdf5634904974255/docs/binomthinultimate.html" target="_blank">9d09fc0</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-03-21
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DongyueXie/misc/blob/c7baf57478b5d87b41274f9f30e4afc7a66cc88f/analysis/binomthinultimate.Rmd" target="_blank">c7baf57</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-03-21
</td>
<td>
wflow_publish(“analysis/binomthinultimate.Rmd”)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>I observe that for a real single cell RNA-Seq dataset, if applying binomial thinning, taking log1p, and fitting linear models, then effects with larger standard error tend to be larger. This is seen from the fact that setting <span class="math inline">\(\alpha=1\)</span> in <code>ash</code> gives higher likelihood than setting <span class="math inline">\(\alpha=0\)</span>. I want to figure out why this happens.</p>
<p>Binomial thinning: given observed counts <span class="math inline">\(n\)</span> , effect size <span class="math inline">\(\beta\)</span> and two groups, draw <span class="math inline">\(y\)</span> either from Binomial<span class="math inline">\((n, \frac{\exp(\beta)}{1\exp(\beta)})\)</span> or Binomial<span class="math inline">\((n, \frac{1}{1\exp(\beta)})\)</span>, depending on group. Without loss of generality, we assign <span class="math inline">\(p = \exp(\beta)}{1\exp(\beta)\)</span> to group one and <span class="math inline">\(1-p\)</span> to group two.</p>
<p>How I tackle this: The standard error of estimated effects is proportional to square root residual variance. Hence, I need to understand the relationship between effects and residual variance. Residual variance is basically <span class="math inline">\(var(\log(y+s))\)</span>, where <span class="math inline">\(s\)</span> is usually taken to be 0.5 or 1.</p>
<p>Of course, different counts and thinning probabilities result in heterogeneity so a linear model with constant variance is not strictly adequate. Most RUV methods do not consider this heterogeneity. How about voom-ruv-limma-ash combination?</p>
<p>Observed counts vary from 0 to a few hundreds of thousands.</p>
<p>By doing theoretical analysis via Taylor series expansion and extensive simulations, my conclusion is that</p>
<ol style="list-style-type: decimal">
<li>For group one with <span class="math inline">\(p=\frac{\exp\beta}{\exp\beta+1}\)</span>, <span class="math inline">\(var(\log(y+s))\)</span> is a unimodal bell shaped function of <span class="math inline">\(\beta\)</span> given <span class="math inline">\(n\)</span> and <span class="math inline">\(s\)</span>. Based on first order Taylor series expansion, variance of <span class="math inline">\(\log(y+s)\)</span> where <span class="math inline">\(y\sim Binomial(n,\frac{\exp\beta}{1+\exp\beta})\)</span> is <span class="math inline">\(\frac{n\exp\beta}{((n+s)\exp\beta+s)^2}\)</span>, which is maximized at <span class="math inline">\(\log(\frac{s}{n+s})\)</span>. Let’s see some plots for true variance function based on simulating <span class="math inline">\(10^5\)</span> binomial samples.</li>
</ol>
<p>Plots on the left in figure below show variance of <span class="math inline">\(\log(y+0.5)\)</span> vs <span class="math inline">\(\beta\)</span> relationships for <span class="math inline">\(n=1,5,10,50\)</span>. The ones in the middle change <span class="math inline">\(\beta\)</span> to <span class="math inline">\(p\)</span> on x axis. The y axis in right plots are sum of <span class="math inline">\(var(\log(y+0.5))\)</span> of two groups, i.e. <span class="math inline">\(var(\log(y_1)+0.5)+var(\log(y_2)+0.5)\)</span>, where <span class="math inline">\(y_1\sim Binomial(n,p)\)</span> and <span class="math inline">\(y_2\sim Binomial(n,p)\)</span>. The idea of these plots are that suppose we have two samples with the same observed counts <span class="math inline">\(n\)</span> and we assign them to two different groups, we want to understand how total variances(proportional to residual variance) vary with <span class="math inline">\(\beta\)</span>?</p>
<p>Plots on the left show that as <span class="math inline">\(n\)</span> goes larger, the <span class="math inline">\(\beta\)</span> maximizing the curve becomes smaller, which conforms to the Taylor series expansion analysis.</p>
<p>The plots in the middle show that the <span class="math inline">\(n=1\)</span> curve is similar to a binomial distribution curve(<span class="math inline">\(p\)</span> vs <span class="math inline">\(np(1-p)\)</span>) while the other ones skew to right.</p>
<pre class="r"><code>library(Biobase)
par(mfrow = c(4,3))

b = seq(-5,5,length.out = 101)
n_list = c(1,5,10,50)
s=0.5

set.seed(12345)

for(n in n_list){
  true_var1 = c()
  true_var2 = c()
  p1 = c()
  p2 = c()
  for(i in 1:length(b)){
    p1[i] = exp(b[i])/(1+exp(b[i]))
    p2[i] = 1/(1+exp(b[i]))
    true_var1 = c(true_var1,var(log(rbinom(1e5,n,p1[i])+s)))
    true_var2 = c(true_var2,var(log(rbinom(1e5,n,p2[i])+s)))
  }
  
  #print(p1[which.max(true_var1)]*n)
  
  plot(b,true_var1,type=&#39;l&#39;,xlab=&#39;beta&#39;,ylab=&#39;variance&#39;,main = paste(&#39;n = &#39;,n))
  abline(v = 0, lty = 2)
  
  plot(n*p1,true_var1,type=&#39;l&#39;,xlab=&#39;p&#39;,ylab=&#39;variance&#39;,main = paste(&#39;n = &#39;,n))
  abline(v = 0, lty = 2)
  
  plot(b,true_var1 + true_var2,type=&#39;l&#39;,xlab=&#39;beta&#39;,ylab=&#39;variance&#39;,main = paste(&#39;n = &#39;,n, &#39;, sum of two groups&#39;))
  abline(v = 0, lty = 2)
  #plot(p,true_var,type=&#39;l&#39;)
  # first order taylor series expansion approximated one
  #lines(b, n*exp(b)*(((2*n+2*s+1)*exp(b)+2*s-1)^2+2*(n-1)*exp(b)) / (4*((n+s)*exp(b)+s)^4),col=4)
}</code></pre>
<p><img src="figure/binomthinultimate.Rmd/unnamed-chunk-2-1.png" width="864" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-2-1">
Past versions of unnamed-chunk-2-1.png
</button>
</p>
<div id="fig-unnamed-chunk-2-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/DongyueXie/misc/blob/7177fb223462fb9ca44ae519f0efc54467adea8c/docs/figure/binomthinultimate.Rmd/unnamed-chunk-2-1.png" target="_blank">7177fb2</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>If we consider <span class="math inline">\(|\beta|\leq 2.5\)</span>, then roughly when <span class="math inline">\(n\geq 15\)</span>, <span class="math inline">\(s.e.(\hat\beta)\)</span> should be monotonically increasing with <span class="math inline">\(|\beta|\)</span>.</p>
<pre class="r"><code>load(&quot;data/scde/scCDT.RData&quot;)
CDT = as.matrix(CDT)
remove.idx = which(rowSums(CDT&gt;0)&lt;=20)
CDT = CDT[-remove.idx,]

#&#39;@param Z count matrix, sample by features
#&#39;@param x 1 for group 1, 0 for group 2
#&#39;@param beta effect of fearures,  0 for null.
#&#39;@return W, thinned matrix
bi_thin = function(Z,x,beta){

  n=nrow(Z)
  p=ncol(Z)
  
  # group index
  g1 = which(x==1)
  g2 = which(x==0)
  
  
  p2 = 1/(1+exp(beta))
  p1 = 1-p2
  P = matrix(nrow = n,ncol = p)
  P[g1,] = t(replicate(length(g1),p1))
  P[g2,] = t(replicate(length(g2),p2))
  
  W = matrix(rbinom(n*p,Z,P),nrow=n)
  
  W
}


set.seed(12345)

#beta = sample(seq(-2.5,2.5,length.out = nrow(CDT)))

beta = rnorm(nrow(CDT),0,0.8)

null_idx = sample(1:nrow(CDT),round(0.5*nrow(CDT)))

beta[null_idx] = 0

which_null = 1*(beta==0)

x = rep(0,ncol(CDT))

x[sample(1:ncol(CDT),round(ncol(CDT)/2))] = 1

Y = bi_thin(t(CDT),x,beta)

#Y = Y[,-which(colSums(Y!=0)&lt;10)]

X = model.matrix(~x)

lmout = limma::lmFit(object = t(log(Y+0.5)), design = X)
#eout  &lt;- limma::eBayes(lmout)
out           = list()
out$betahat   = lmout$coefficients[, 2]
out$sebetahat = lmout$stdev.unscaled[, 2] * lmout$sigma
#out$pvalues   = eout$p.value[, 2]

par(mfrow=c(2,2))

plot(beta,out$betahat,ylab=&#39;beta_hat&#39;)
abline(a=0,b=1)
#plot(abs(beta),out$sebetahat,ylab = &#39;s.e.&#39;)
plot(abs(out$betahat),out$sebetahat,xlab = &#39;abs(beta_hat)&#39;,ylab = &#39;s.e.&#39;)

alphas = seq(0,1,length.out = 11)

loglik = c()
auc = c()

for(alpha in alphas){
  ash_ash = ashr::ash(out$betahat,out$sebetahat,alpha=alpha)
  loglik = c(loglik,ash_ash$loglik)
  auc = c(auc,pROC::roc(which_null,ash_ash$result$lfsr)$auc)
}

plot(alphas,loglik,type=&#39;l&#39;,xlab = &#39;alpha&#39;)
plot(alphas,auc,type=&#39;l&#39;,xlab = &#39;alpha&#39;)</code></pre>
<p><img src="figure/binomthinultimate.Rmd/unnamed-chunk-3-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-3-1">
Past versions of unnamed-chunk-3-1.png
</button>
</p>
<div id="fig-unnamed-chunk-3-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/DongyueXie/misc/blob/7177fb223462fb9ca44ae519f0efc54467adea8c/docs/figure/binomthinultimate.Rmd/unnamed-chunk-3-1.png" target="_blank">7177fb2</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Another run:</p>
<pre class="r"><code>#beta = sample(seq(-2.5,2.5,length.out = nrow(CDT)))

beta = rnorm(nrow(CDT),0,0.8)

null_idx = sample(1:nrow(CDT),round(0.5*nrow(CDT)))

beta[null_idx] = 0

which_null = 1*(beta==0)

x = rep(0,ncol(CDT))

x[sample(1:ncol(CDT),round(ncol(CDT)/2))] = 1

Y = bi_thin(t(CDT),x,beta)

#Y = Y[,-which(colSums(Y!=0)&lt;10)]

X = model.matrix(~x)

lmout = limma::lmFit(object = t(log(Y+0.5)), design = X)
#eout  &lt;- limma::eBayes(lmout)
out           = list()
out$betahat   = lmout$coefficients[, 2]
out$sebetahat = lmout$stdev.unscaled[, 2] * lmout$sigma
#out$pvalues   = eout$p.value[, 2]

par(mfrow=c(2,2))

plot(beta,out$betahat,ylab=&#39;beta_hat&#39;)
abline(a=0,b=1)
#plot(abs(beta),out$sebetahat,ylab = &#39;s.e.&#39;)
plot(abs(out$betahat),out$sebetahat,xlab = &#39;abs(beta_hat)&#39;,ylab = &#39;s.e.&#39;)

alphas = seq(0,1,length.out = 11)

loglik = c()
auc = c()

for(alpha in alphas){
  ash_ash = ashr::ash(out$betahat,out$sebetahat,alpha=alpha)
  loglik = c(loglik,ash_ash$loglik)
  auc = c(auc,pROC::roc(which_null,ash_ash$result$lfsr)$auc)
}

plot(alphas,loglik,type=&#39;l&#39;,xlab = &#39;alpha&#39;)
plot(alphas,auc,type=&#39;l&#39;,xlab = &#39;alpha&#39;)</code></pre>
<p><img src="figure/binomthinultimate.Rmd/unnamed-chunk-4-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-4-1">
Past versions of unnamed-chunk-4-1.png
</button>
</p>
<div id="fig-unnamed-chunk-4-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/DongyueXie/misc/blob/7177fb223462fb9ca44ae519f0efc54467adea8c/docs/figure/binomthinultimate.Rmd/unnamed-chunk-4-1.png" target="_blank">7177fb2</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The plot of <span class="math inline">\(\beta\)</span> vs <span class="math inline">\(\hat\beta\)</span> shows that if <span class="math inline">\(\hat\beta\)</span> is large, then <span class="math inline">\(\beta\)</span> is large. But if <span class="math inline">\(\beta\)</span> is large, <span class="math inline">\(\hat\beta\)</span> is not necessarily large because counts are small, the most we can thin is limited. In this sense, some of the ‘true’ <span class="math inline">\(\beta\)</span>s used to thin the counts are not really ‘true’ since data do not contain such information.</p>
<p>As long as <span class="math inline">\(\hat\beta\)</span> is a consistent estimator, and sample size is large(701 here, large rnough), <span class="math inline">\(\hat\beta\)</span> should be close enough to <span class="math inline">\(\beta\)</span>.</p>
<p>However, it seems that the Taylor series and log Binomial analysis still cannot explain the <span class="math inline">\(\alpha\)</span> stuff. Becuase according to above analysis, since most genes have small counts, then effects with larger s.e. should be smaller. This is certainly not true here. On the other hand, if we only consider genes with maximum counts less or equal to 1, then apply binomial thinning, the results are:</p>
<pre class="r"><code>ii = which(rowMax(CDT)&lt;=1)

set.seed(12345)
par(mfrow=c(2,2))

#beta = sample(seq(-2.5,2.5,length.out = nrow(CDT)))

beta = rnorm(nrow(CDT[ii,]),0,0.8)

null_idx = sample(1:nrow(CDT[ii,]),round(0.5*nrow(CDT[ii,])))

beta[null_idx] = 0

which_null = 1*(beta==0)

x = rep(0,ncol(CDT[ii,]))

x[sample(1:ncol(CDT[ii,]),round(ncol(CDT[ii,])/2))] = 1

Y = bi_thin(t(CDT[ii,]),x,beta)

#Y = Y[,-which(colSums(Y!=0)&lt;10)]

X = model.matrix(~x)

lmout = limma::lmFit(object = t(log(Y+0.5)), design = X)
#eout  &lt;- limma::eBayes(lmout)
out           = list()
out$betahat   = lmout$coefficients[, 2]
out$sebetahat = lmout$stdev.unscaled[, 2] * lmout$sigma
#out$pvalues   = eout$p.value[, 2]

plot(beta,out$betahat,ylab=&#39;beta_hat&#39;)
abline(a=0,b=1)
#plot(abs(beta),out$sebetahat)
plot(abs(out$betahat),out$sebetahat,pch = 18,xlab = &#39;abs(beta_hat)&#39;,ylab = &#39;s.e.&#39;)

alphas = seq(0,1,length.out = 11)

loglik = c()
auc = c()

for(alpha in alphas){
  ash_ash = ashr::ash(out$betahat,out$sebetahat,alpha=alpha)
  loglik = c(loglik,ash_ash$loglik)
  auc = c(auc,pROC::roc(which_null,ash_ash$result$lfsr)$auc)
}

plot(alphas,loglik,type=&#39;l&#39;)
plot(alphas,auc,type=&#39;l&#39;)</code></pre>
<p><img src="figure/binomthinultimate.Rmd/unnamed-chunk-5-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-5-1">
Past versions of unnamed-chunk-5-1.png
</button>
</p>
<div id="fig-unnamed-chunk-5-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/DongyueXie/misc/blob/7177fb223462fb9ca44ae519f0efc54467adea8c/docs/figure/binomthinultimate.Rmd/unnamed-chunk-5-1.png" target="_blank">7177fb2</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Another run</p>
<pre class="r"><code>par(mfrow=c(2,2))
beta = rnorm(nrow(CDT[ii,]),0,0.8)

null_idx = sample(1:nrow(CDT[ii,]),round(0.5*nrow(CDT[ii,])))

beta[null_idx] = 0

which_null = 1*(beta==0)

x = rep(0,ncol(CDT[ii,]))

x[sample(1:ncol(CDT[ii,]),round(ncol(CDT[ii,])/2))] = 1

Y = bi_thin(t(CDT[ii,]),x,beta)

#Y = Y[,-which(colSums(Y!=0)&lt;10)]

X = model.matrix(~x)

lmout = limma::lmFit(object = t(log(Y+0.5)), design = X)
#eout  &lt;- limma::eBayes(lmout)
out           = list()
out$betahat   = lmout$coefficients[, 2]
out$sebetahat = lmout$stdev.unscaled[, 2] * lmout$sigma
#out$pvalues   = eout$p.value[, 2]

plot(beta,out$betahat,ylab=&#39;beta_hat&#39;)
abline(a=0,b=1)
#plot(abs(beta),out$sebetahat)
plot(abs(out$betahat),out$sebetahat,pch = 18,xlab = &#39;abs(beta_hat)&#39;,ylab = &#39;s.e.&#39;)

alphas = seq(0,1,length.out = 11)

loglik = c()
auc = c()

for(alpha in alphas){
  ash_ash = ashr::ash(out$betahat,out$sebetahat,alpha=alpha)
  loglik = c(loglik,ash_ash$loglik)
  auc = c(auc,pROC::roc(which_null,ash_ash$result$lfsr)$auc)
}

plot(alphas,loglik,type=&#39;l&#39;)
plot(alphas,auc,type=&#39;l&#39;)</code></pre>
<p><img src="figure/binomthinultimate.Rmd/unnamed-chunk-6-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-6-1">
Past versions of unnamed-chunk-6-1.png
</button>
</p>
<div id="fig-unnamed-chunk-6-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/DongyueXie/misc/blob/7177fb223462fb9ca44ae519f0efc54467adea8c/docs/figure/binomthinultimate.Rmd/unnamed-chunk-6-1.png" target="_blank">7177fb2</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>If we only consider genes with maximum counts greater or equal to 50, then apply binomial thinning, the results are:</p>
<pre class="r"><code>par(mfrow=c(2,2))
ii = which(rowMax(CDT)&gt;=50)

set.seed(12345)

#beta = sample(seq(-2.5,2.5,length.out = nrow(CDT)))

beta = rnorm(nrow(CDT[ii,]),0,0.8)

null_idx = sample(1:nrow(CDT[ii,]),round(0.5*nrow(CDT[ii,])))

beta[null_idx] = 0

which_null = 1*(beta==0)

x = rep(0,ncol(CDT[ii,]))

x[sample(1:ncol(CDT[ii,]),round(ncol(CDT[ii,])/2))] = 1

Y = bi_thin(t(CDT[ii,]),x,beta)

#Y = Y[,-which(colSums(Y!=0)&lt;10)]

X = model.matrix(~x)

lmout = limma::lmFit(object = t(log(Y+0.5)), design = X)
#eout  &lt;- limma::eBayes(lmout)
out           = list()
out$betahat   = lmout$coefficients[, 2]
out$sebetahat = lmout$stdev.unscaled[, 2] * lmout$sigma
#out$pvalues   = eout$p.value[, 2]

plot(beta,out$betahat,ylab=&#39;beta_hat&#39;)
abline(a=0,b=1)
#plot(abs(beta),out$sebetahat)
plot(abs(out$betahat),out$sebetahat,pch = 18,xlab = &#39;abs(beta_hat)&#39;,ylab = &#39;s.e.&#39;)

alphas = seq(0,1,length.out = 11)

loglik = c()
auc = c()

for(alpha in alphas){
  ash_ash = ashr::ash(out$betahat,out$sebetahat,alpha=alpha)
  loglik = c(loglik,ash_ash$loglik)
  auc = c(auc,pROC::roc(which_null,ash_ash$result$lfsr)$auc)
}

plot(alphas,loglik,type=&#39;l&#39;)
plot(alphas,auc,type=&#39;l&#39;)</code></pre>
<p><img src="figure/binomthinultimate.Rmd/unnamed-chunk-7-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-7-1">
Past versions of unnamed-chunk-7-1.png
</button>
</p>
<div id="fig-unnamed-chunk-7-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/DongyueXie/misc/blob/7177fb223462fb9ca44ae519f0efc54467adea8c/docs/figure/binomthinultimate.Rmd/unnamed-chunk-7-1.png" target="_blank">7177fb2</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Another run</p>
<pre class="r"><code>par(mfrow=c(2,2))
beta = rnorm(nrow(CDT[ii,]),0,0.8)

null_idx = sample(1:nrow(CDT[ii,]),round(0.5*nrow(CDT[ii,])))

beta[null_idx] = 0

which_null = 1*(beta==0)

x = rep(0,ncol(CDT[ii,]))

x[sample(1:ncol(CDT[ii,]),round(ncol(CDT[ii,])/2))] = 1

Y = bi_thin(t(CDT[ii,]),x,beta)

#Y = Y[,-which(colSums(Y!=0)&lt;10)]

X = model.matrix(~x)

lmout = limma::lmFit(object = t(log(Y+0.5)), design = X)
#eout  &lt;- limma::eBayes(lmout)
out           = list()
out$betahat   = lmout$coefficients[, 2]
out$sebetahat = lmout$stdev.unscaled[, 2] * lmout$sigma
#out$pvalues   = eout$p.value[, 2]

plot(beta,out$betahat,ylab=&#39;beta_hat&#39;)
abline(a=0,b=1)
#plot(abs(beta),out$sebetahat)
plot(abs(out$betahat),out$sebetahat,pch = 18,xlab = &#39;abs(beta_hat)&#39;,ylab = &#39;s.e.&#39;)

alphas = seq(0,1,length.out = 11)

loglik = c()
auc = c()

for(alpha in alphas){
  ash_ash = ashr::ash(out$betahat,out$sebetahat,alpha=alpha)
  loglik = c(loglik,ash_ash$loglik)
  auc = c(auc,pROC::roc(which_null,ash_ash$result$lfsr)$auc)
}

plot(alphas,loglik,type=&#39;l&#39;)
plot(alphas,auc,type=&#39;l&#39;)</code></pre>
<p><img src="figure/binomthinultimate.Rmd/unnamed-chunk-8-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-8-1">
Past versions of unnamed-chunk-8-1.png
</button>
</p>
<div id="fig-unnamed-chunk-8-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/DongyueXie/misc/blob/7177fb223462fb9ca44ae519f0efc54467adea8c/docs/figure/binomthinultimate.Rmd/unnamed-chunk-8-1.png" target="_blank">7177fb2</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>So there must have something to do with the 0’s.</p>
<p>Let’s directly work with <span class="math inline">\(\hat{\sigma}^2\)</span>,</p>
<span class="math display">\[\begin{equation}
    \hat{\sigma}^2 = \frac{1}{n-2}(\sum_{i\in Group1}(y_i-\bar{y}_1)^2+\sum_{i\in Group2}(y_i-\bar{y}_2)^2),
\end{equation}\]</span>
<p>where <span class="math inline">\(\bar{y}_1\)</span> is the sample mean of group 1. Let <span class="math inline">\(n_1\)</span> be the number of samples in group, <span class="math inline">\(p_1^0\)</span> be the proportion of zeros or <span class="math inline">\(\log(s)\)</span> in group 1, and <span class="math inline">\(y_0\)</span> denote zero or <span class="math inline">\(\log(s)\)</span>, then</p>
<span class="math display">\[\begin{equation}
    \begin{split}
        \sum_{i\in Group1}(y_i-\bar{y}_1)^2 &amp;= \sum_{\substack{i\in Group1 \\ y_i\neq y_0}}(y_i-\bar{y}_1)^2 + n_1p_1^0(y_0-\bar{y}_1)^2
        \\&amp;= \sum_{\substack{i\in Group1 \\ y_i\neq y_0}} y_i ^2 - 2p_1^0y_0\sum_{\substack{i\in Group1 \\ y_i\neq y_0}} y_i - \frac{1}{n_1}(\sum_{\substack{i\in Group1 \\ y_i\neq y_0}} y_i)^2 + n_1 p_1^0(1-p_1^0)y_0^2
        \\&amp;= n_1(1-p_1^0)\text{Var}\{y_i,\in Group1, y_i\neq y_0\} + \frac{(\sum_{\substack{i\in Group1 \\ y_i\neq y_0}} y_i)^2}{n_1/p_1^0-n_1} + n_1 p_1^0(1-p_1^0)y_0(y_0-2\bar{y}_{1,y_i\neq y_0})
    \end{split}
\end{equation}\]</span>
<pre class="r"><code>G=1000
n=100
Y = matrix(0,nrow=G,ncol=n)
Y = t(Y)
Y[1:10,] = 1
Y[51:60,] = 1
Y = t(Y)

set.seed(12345)
beta = rnorm(G,0,0.8)

null_idx = sample(1:G,round(0.5*G))

beta[null_idx] = 0

which_null = 1*(beta==0)

x = rep(0,n)

x[sample(1:n,round(n/2))] = 1

Y = bi_thin(t(Y),x,beta)

#Y = Y[,-which(colSums(Y!=0)&lt;10)]

X = model.matrix(~x)

lmout = limma::lmFit(object = t(log(Y+0.5)), design = X)
#eout  &lt;- limma::eBayes(lmout)
out           = list()
out$betahat   = lmout$coefficients[, 2]
out$sebetahat = lmout$stdev.unscaled[, 2] * lmout$sigma
#out$pvalues   = eout$p.value[, 2]

plot(beta,out$betahat,ylab=&#39;beta_hat&#39;)
abline(a=0,b=1)
#plot(abs(beta),out$sebetahat)
plot((out$betahat),out$sebetahat,pch = 18,xlab = &#39;(beta_hat)&#39;,ylab = &#39;s.e.&#39;)

alphas = seq(0,1,length.out = 11)

loglik = c()
auc = c()

for(alpha in alphas){
  ash_ash = ashr::ash(out$betahat,out$sebetahat,alpha=alpha)
  loglik = c(loglik,ash_ash$loglik)
  auc = c(auc,pROC::roc(which_null,ash_ash$result$lfsr)$auc)
}

plot(alphas,loglik,type=&#39;l&#39;)
plot(alphas,auc,type=&#39;l&#39;)</code></pre>
<pre class="r"><code>G=1000
n=100
Y = matrix(0,nrow=G,ncol=n)
Y = t(Y)
Y[1:5,] = 1
Y[6:10,] = 50
Y[51:55,] = 1
Y[56:60,] = 1
Y = t(Y)

set.seed(12345)
beta = rnorm(G,0,0.8)

null_idx = sample(1:G,round(0.5*G))

beta[null_idx] = 0

which_null = 1*(beta==0)

x = rep(0,n)

x[sample(1:n,round(n/2))] = 1

Y = bi_thin(t(Y),x,beta)

#Y = Y[,-which(colSums(Y!=0)&lt;10)]

X = model.matrix(~x)

lmout = limma::lmFit(object = t(log(Y+0.5)), design = X)
#eout  &lt;- limma::eBayes(lmout)
out           = list()
out$betahat   = lmout$coefficients[, 2]
out$sebetahat = lmout$stdev.unscaled[, 2] * lmout$sigma
#out$pvalues   = eout$p.value[, 2]

plot(beta,out$betahat,ylab=&#39;beta_hat&#39;)
abline(a=0,b=1)
#plot(abs(beta),out$sebetahat)
plot((out$betahat),out$sebetahat,pch = 18,xlab = &#39;(beta_hat)&#39;,ylab = &#39;s.e.&#39;)

alphas = seq(0,1,length.out = 11)

loglik = c()
auc = c()

for(alpha in alphas){
  ash_ash = ashr::ash(out$betahat,out$sebetahat,alpha=alpha)
  loglik = c(loglik,ash_ash$loglik)
  auc = c(auc,pROC::roc(which_null,ash_ash$result$lfsr)$auc)
}

plot(alphas,loglik,type=&#39;l&#39;)
plot(alphas,auc,type=&#39;l&#39;)</code></pre>
<div id="earlier-analysis" class="section level1">
<h1>Earlier analysis</h1>
<div id="taylor-series-approximation" class="section level2">
<h2>Taylor series approximation</h2>
<p>First order</p>
<pre class="r"><code>mean_log = function(n,p,s){
  log(n*p+s)
}

var_log = function(n,p,s){
  n*p*(1-p)/(n*p+s)^2 
}




s = 1


  
  set.seed(12345)
  
  for(p in c(0.1,0.5,0.9)){
  
  true_var = c()
  approx_var = c()
  
  true_mean = c()
  approx_mean = c()
  
  n_list = 0:ceiling(20/p)
  for(n in n_list){
    y = rbinom(1e5,n,p)
    true_var = c(true_var,var(log(y+s)))
    true_mean = c(true_mean,mean(log(y+s)))
    approx_var = c(approx_var,var_log(n,p,s))
    approx_mean = c(approx_mean,mean_log(n,p,s))
  }
  
  par(mfrow = c(1,2))
  
  plot(n_list*p,true_mean,type=&#39;l&#39;,col=2,xlab = &#39;n*p&#39;,ylab = &#39;mean&#39;,ylim = range(c(true_mean,approx_mean)),main = paste(&#39;p =&#39;,p))
  lines(n_list*p,approx_mean,col=4)
  legend(&#39;bottomright&#39;,c(&#39;true&#39;,&#39;approx&#39;),lty=c(1,1),col=c(2,4))
  
  plot(n_list*p,true_var,type=&#39;l&#39;,col=2,xlab = &#39;n*p&#39;,ylab = &#39;varaince&#39;,ylim = range(c(true_var,approx_var)),main = paste(&#39;p =&#39;,p))
  lines(n_list*p,approx_var,col=4)
  legend(&#39;topright&#39;,c(&#39;true&#39;,&#39;approx&#39;),lty=c(1,1),col=c(2,4))

}</code></pre>
<p><img src="figure/binomthinultimate.Rmd/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-11-1">
Past versions of unnamed-chunk-11-1.png
</button>
</p>
<div id="fig-unnamed-chunk-11-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/DongyueXie/misc/blob/7177fb223462fb9ca44ae519f0efc54467adea8c/docs/figure/binomthinultimate.Rmd/unnamed-chunk-11-1.png" target="_blank">7177fb2</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/binomthinultimate.Rmd/unnamed-chunk-11-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-11-2">
Past versions of unnamed-chunk-11-2.png
</button>
</p>
<div id="fig-unnamed-chunk-11-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/DongyueXie/misc/blob/7177fb223462fb9ca44ae519f0efc54467adea8c/docs/figure/binomthinultimate.Rmd/unnamed-chunk-11-2.png" target="_blank">7177fb2</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/binomthinultimate.Rmd/unnamed-chunk-11-3.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-11-3">
Past versions of unnamed-chunk-11-3.png
</button>
</p>
<div id="fig-unnamed-chunk-11-3" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/DongyueXie/misc/blob/7177fb223462fb9ca44ae519f0efc54467adea8c/docs/figure/binomthinultimate.Rmd/unnamed-chunk-11-3.png" target="_blank">7177fb2</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Second order</p>
<pre class="r"><code>mean_log = function(n,p,s){
  log(n*p+s)-n*p*(1-p)/(2*(n*p+s)^2)
}

var_log = function(n,p,s){
  n*p*(1-p)/(n*p+s)^2 + n*p*(1-p)*(1+(2*n-6)*p*(1-p))/(4*(n*p+s)^4) - n*p*(1-p)*(1-2*p)/(n*p+s)^3
}




s = 0.5


  
  set.seed(12345)
  
  for(p in c(0.1,0.5,0.9)){
  
  true_var = c()
  approx_var = c()
  
  true_mean = c()
  approx_mean = c()
  
  n_list = 0:ceiling(20/p)
  for(n in n_list){
    y = rbinom(1e5,n,p)
    true_var = c(true_var,var(log(y+s)))
    true_mean = c(true_mean,mean(log(y+s)))
    approx_var = c(approx_var,var_log(n,p,s))
    approx_mean = c(approx_mean,mean_log(n,p,s))
  }
  
  par(mfrow = c(1,2))
  
  plot(n_list*p,true_mean,type=&#39;l&#39;,col=2,xlab = &#39;n*p&#39;,ylab = &#39;mean&#39;,ylim = range(c(true_mean,approx_mean)),main = paste(&#39;p =&#39;,p))
  lines(n_list*p,approx_mean,col=4)
  legend(&#39;bottomright&#39;,c(&#39;true&#39;,&#39;approx&#39;),lty=c(1,1),col=c(2,4))
  
  plot(n_list*p,true_var,type=&#39;l&#39;,col=2,xlab = &#39;n*p&#39;,ylab = &#39;varaince&#39;,ylim = range(c(true_var,approx_var)),main = paste(&#39;p =&#39;,p))
  lines(n_list*p,approx_var,col=4)
  legend(&#39;topright&#39;,c(&#39;true&#39;,&#39;approx&#39;),lty=c(1,1),col=c(2,4))

}</code></pre>
<p><img src="figure/binomthinultimate.Rmd/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-12-1">
Past versions of unnamed-chunk-12-1.png
</button>
</p>
<div id="fig-unnamed-chunk-12-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/DongyueXie/misc/blob/7177fb223462fb9ca44ae519f0efc54467adea8c/docs/figure/binomthinultimate.Rmd/unnamed-chunk-12-1.png" target="_blank">7177fb2</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/binomthinultimate.Rmd/unnamed-chunk-12-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-12-2">
Past versions of unnamed-chunk-12-2.png
</button>
</p>
<div id="fig-unnamed-chunk-12-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/DongyueXie/misc/blob/7177fb223462fb9ca44ae519f0efc54467adea8c/docs/figure/binomthinultimate.Rmd/unnamed-chunk-12-2.png" target="_blank">7177fb2</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/binomthinultimate.Rmd/unnamed-chunk-12-3.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-12-3">
Past versions of unnamed-chunk-12-3.png
</button>
</p>
<div id="fig-unnamed-chunk-12-3" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/DongyueXie/misc/blob/7177fb223462fb9ca44ae519f0efc54467adea8c/docs/figure/binomthinultimate.Rmd/unnamed-chunk-12-3.png" target="_blank">7177fb2</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="large-np" class="section level2">
<h2>Large <span class="math inline">\(np\)</span></h2>
<pre class="r"><code>set.seed(12345)
p_list = seq(0.1,0.5,length.out = 100)

n = 50
s = 0.5


true_var = c()
approx_var = c()
  
for(p in p_list){
  y1 = rbinom(1e5,n,p)
  y2 = rbinom(1e5,n,1-p)
  true_var = c(true_var,var(log(y1+s))+var(log(y2+s)))
  approx_var = c(approx_var,var_log(n,p,s)+var_log(n,1-p,s))
}

par(mfrow=c(1,1))

plot(log(1/p_list-1),true_var,xlab = &#39;beta&#39;, ylab= &#39;variance&#39;,type=&#39;l&#39;,col=2,ylim = range(c(true_var,approx_var)))
lines(log(1/p_list-1),approx_var,col=4)
legend(&#39;topleft&#39;,c(&#39;true&#39;,&#39;approx&#39;),lty=c(1,1),col=c(2,4))</code></pre>
<p><img src="figure/binomthinultimate.Rmd/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-13-1">
Past versions of unnamed-chunk-13-1.png
</button>
</p>
<div id="fig-unnamed-chunk-13-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/DongyueXie/misc/blob/7177fb223462fb9ca44ae519f0efc54467adea8c/docs/figure/binomthinultimate.Rmd/unnamed-chunk-13-1.png" target="_blank">7177fb2</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="small-np" class="section level2">
<h2>Small <span class="math inline">\(np\)</span></h2>
<pre class="r"><code>set.seed(12345)
p_list = seq(0.1,0.5,length.out = 100)

n = 5
s = 0.5


true_var = c()
approx_var = c()
  
for(p in p_list){
  y1 = rbinom(1e5,n,p)
  y2 = rbinom(1e5,n,1-p)
  true_var = c(true_var,var(log(y1+s))+var(log(y2+s)))
  approx_var = c(approx_var,var_log(n,p,s)+var_log(n,1-p,s))
}

par(mfrow=c(1,1))

plot(log(1/p_list-1),true_var,xlab = &#39;beta&#39;, ylab= &#39;variance&#39;,type=&#39;l&#39;,col=2,ylim = range(c(true_var,approx_var)))
lines(log(1/p_list-1),approx_var,col=4)
legend(&#39;topright&#39;,c(&#39;true&#39;,&#39;approx&#39;),lty=c(1,1),col=c(2,4))</code></pre>
<p><img src="figure/binomthinultimate.Rmd/unnamed-chunk-14-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-14-1">
Past versions of unnamed-chunk-14-1.png
</button>
</p>
<div id="fig-unnamed-chunk-14-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/DongyueXie/misc/blob/7177fb223462fb9ca44ae519f0efc54467adea8c/docs/figure/binomthinultimate.Rmd/unnamed-chunk-14-1.png" target="_blank">7177fb2</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>set.seed(12345)
n_list = 0:100


s = 0.5
p = 0.1

true_var = c()
approx_var = c()
  
for(n in n_list){
  y1 = rbinom(1e5,n,p)
  y2 = rbinom(1e5,n,1-p)
  true_var = c(true_var,var(log(y1+s))+var(log(y2+s)))
  approx_var = c(approx_var,var_log(n,p,s)+var_log(n,1-p,s))
}

plot(n_list,true_var,xlab = &#39;n&#39;, ylab= &#39;variance&#39;,type=&#39;l&#39;,col=2)
lines(n_list,approx_var,col=4)
legend(&#39;topleft&#39;,c(&#39;true&#39;,&#39;approx&#39;),lty=c(1,1),col=c(2,4))</code></pre>
<p><img src="figure/binomthinultimate.Rmd/unnamed-chunk-15-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-15-1">
Past versions of unnamed-chunk-15-1.png
</button>
</p>
<div id="fig-unnamed-chunk-15-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/DongyueXie/misc/blob/7177fb223462fb9ca44ae519f0efc54467adea8c/docs/figure/binomthinultimate.Rmd/unnamed-chunk-15-1.png" target="_blank">7177fb2</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Let’s start with the simplest case: we have a count matrix <span class="math inline">\(Y\)</span>, whose all entries are the same! Then we apply binomial thinning to <span class="math inline">\(Y\)</span>.</p>
<p>In the following simulation, we set the number of samples to be <span class="math inline">\(n=500\)</span>, the number of genes <span class="math inline">\(p=5000\)</span> and entries of <span class="math inline">\(Y\)</span> vary from <span class="math inline">\(100, 50, 30, 20, 10, 5, 1.\)</span> Half of samples are from group 1 and the rest are from group 2. About <span class="math inline">\(90\%\)</span> of genes have no effects and the rest have effects <span class="math inline">\(\beta_j\)</span> generated from standard normal distribution.</p>
<p>We fit a simple linear regression to each gene with log transformed counts and compare the log-likelihood from <code>ash</code>, setting <span class="math inline">\(\alpha=0\)</span> and <span class="math inline">\(\alpha=1\)</span>, and plot <span class="math inline">\(\beta\)</span> vs <span class="math inline">\(\hat\beta\)</span> as well as <span class="math inline">\(\beta\)</span> vs <span class="math inline">\(s.e.(\hat\beta)\)</span> for each case.</p>
<pre class="r"><code>#&#39;@param Z count matrix, sample by features
#&#39;@param x 1 for group 1, 0 for group 2
#&#39;@param beta effect of fearures,  0 for null.
#&#39;@return W, thinned matrix
bi_thin = function(Z,x,beta){
  
  n=nrow(Z)
  p=ncol(Z)
  
  # group index
  g1 = which(x==1)
  g2 = which(x==0)
  
  
  p2 = 1/(1+exp(beta))
  p1 = 1-p2
  P = matrix(nrow = n,ncol = p)
  P[g1,] = t(replicate(length(g1),p1))
  P[g2,] = t(replicate(length(g2),p2))
  
  W = matrix(rbinom(n*p,Z,P),nrow=n)
  
  W
}

library(limma)

n=500
p=5000
set.seed(12345)

loglik0 = c()
auc0 = c()
loglik1 = c()
auc1 = c()

par(mfrow = c(1,2))
for(ni in c(100,50,30,20,10,5,1)){
  Y = matrix(rep(ni,n*p),nrow=p,ncol=n)

  group_idx = c(rep(1,n/2),rep(0,n/2))
  X = model.matrix(~group_idx)

  b = seq(1,10,length.out = p)
  b[sample(1:p,0.9*p)] = 0
  which_null = 1*(b==0)

  W = bi_thin(t(Y),group_idx,b)

  lmout &lt;- limma::lmFit(object = t(log(W+0.5)), design = X)

  ash0 = ashr::ash(lmout$coefficients[, 2],lmout$stdev.unscaled[, 2]*lmout$sigma,alpha=0)
  loglik0 = c(loglik0,ash0$loglik)
  auc0 = c(auc0,pROC::roc(which_null,ash0$result$lfsr)$auc)

  ash1 = ashr::ash(lmout$coefficients[, 2],lmout$stdev.unscaled[, 2]*lmout$sigma,alpha=1)
  loglik1 = c(loglik1,ash1$loglik)
  auc1 = c(auc1,pROC::roc(which_null,ash1$result$lfsr)$auc)
  
  
  plot((b),lmout$coefficients[, 2],xlab=&#39;beta&#39;,ylab=&#39;estimated&#39;,col=&#39;grey60&#39;,main=paste(&#39;count:&#39;,ni))
  abline(a=0,b=1)
  plot(abs(b),lmout$stdev.unscaled[, 2]*lmout$sigma,xlab=&#39;absolute value of beta&#39;,ylab=&#39;sd of beta_hat&#39;,main=paste(&#39;count:&#39;,ni))
}

knitr::kable(cbind(c(100,50,30,20,10,5,1),loglik0,loglik1),col.names = c(&#39;Y&#39;,&#39;alpha0&#39;,&#39;alpha1&#39;),
             caption = &#39;log likelihood&#39;)

knitr::kable(cbind(c(100,50,30,20,10,5,1),auc0,auc1),col.names = c(&#39;Y&#39;,&#39;alpha0&#39;,&#39;alpha1&#39;),
             caption = &#39;AUC&#39;)</code></pre>
<p>So when counts in <span class="math inline">\(Y\)</span> are large, setting <span class="math inline">\(\alpha=1\)</span> gives higher likelihood. Estimated <span class="math inline">\(\beta\)</span>s are close to true <span class="math inline">\(\beta\)</span>s and apparently, scale of <span class="math inline">\(\beta\)</span> and <span class="math inline">\(var(\hat\beta)\)</span> are positively correlated. While when counts in <span class="math inline">\(Y\)</span> are small, things are opposite.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.5.1 (2018-07-02)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Scientific Linux 7.4 (Nitrogen)

Matrix products: default
BLAS/LAPACK: /software/openblas-0.2.19-el7-x86_64/lib/libopenblas_haswellp-r0.2.19.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
[1] Matrix_1.2-15       Biobase_2.42.0      BiocGenerics_0.28.0

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.2        pillar_1.3.1      plyr_1.8.4       
 [4] compiler_3.5.1    later_0.7.5       git2r_0.26.1     
 [7] workflowr_1.6.0   iterators_1.0.10  tools_3.5.1      
[10] digest_0.6.18     tibble_2.1.1      gtable_0.2.0     
[13] evaluate_0.12     lattice_0.20-38   pkgconfig_2.0.2  
[16] rlang_0.4.0       foreach_1.4.4     yaml_2.2.0       
[19] dplyr_0.8.0.1     stringr_1.3.1     knitr_1.20       
[22] pROC_1.13.0       fs_1.3.1          tidyselect_0.2.5 
[25] rprojroot_1.3-2   grid_3.5.1        glue_1.3.0       
[28] R6_2.3.0          rmarkdown_1.10    mixsqp_0.2-2     
[31] limma_3.38.2      purrr_0.3.2       ggplot2_3.1.1    
[34] ashr_2.2-39       magrittr_1.5      whisker_0.3-2    
[37] scales_1.0.0      backports_1.1.2   promises_1.0.1   
[40] codetools_0.2-15  htmltools_0.3.6   MASS_7.3-51.1    
[43] assertthat_0.2.0  colorspace_1.3-2  httpuv_1.4.5     
[46] stringi_1.2.4     lazyeval_0.2.1    munsell_0.5.0    
[49] doParallel_1.0.14 pscl_1.5.2        truncnorm_1.0-8  
[52] SQUAREM_2017.10-1 crayon_1.3.4     </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
