<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Dongyue Xie" />

<meta name="date" content="2019-12-06" />

<title>Gene expression, microbiota and genotype</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">misc</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/DongyueXie/misc">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Gene expression, microbiota and genotype</h1>
<h4 class="author">Dongyue Xie</h4>
<h4 class="date">2019-12-06</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-01-21
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>misc/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.5.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20191122code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20191122)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20191122code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20191122)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomDongyueXiemisctree0aa049d9988a7bf1efef74962f687746fcaa3a59targetblank0aa049da"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/DongyueXie/misc/tree/0aa049d9988a7bf1efef74962f687746fcaa3a59" target="_blank">0aa049d</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomDongyueXiemisctree0aa049d9988a7bf1efef74962f687746fcaa3a59targetblank0aa049da" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  analysis/BartXbart.Rmd
    Untracked:  data/ROTS_results.RData
    Untracked:  data/pbmc.rds
    Untracked:  data/pbmc_counts.rds

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DongyueXie/misc/blob/0aa049d9988a7bf1efef74962f687746fcaa3a59/analysis/consulting2.Rmd" target="_blank">0aa049d</a>
</td>
<td>
Dongyue Xie
</td>
<td>
2020-01-21
</td>
<td>
wflow_publish(“analysis/consulting2.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DongyueXie/misc/c9702a5b5e577deee571ec7b69b617405a67bf9c/docs/consulting2.html" target="_blank">c9702a5</a>
</td>
<td>
Dongyue Xie
</td>
<td>
2019-12-07
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DongyueXie/misc/blob/97861d240372a578471634277ec904d0f248e17f/analysis/consulting2.Rmd" target="_blank">97861d2</a>
</td>
<td>
Dongyue Xie
</td>
<td>
2019-12-07
</td>
<td>
wflow_publish(“analysis/consulting2.Rmd”)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="overview" class="section level1">
<h1>Overview</h1>
<p>The experiment is a <span class="math inline">\(2\times 3\)</span> design with two genotypes(cre +, cre -) and three treatments(SPF, Healthy and CMA). Each cell has unequal observations so the design is unbalanced.</p>
<p>The observations are Ct(cycle thresholds), which is defined as the number of cycles required for the fluorescent signal to cross the threshold (ie. exceeds background level). Ct levels are inversely proportional to the amount of target nucleic acid in the sample. Each observation has two numbers, Ct of target gene and Ct of housekeeping gene. The delta Ct, is defined as Ct(target gene) - Ct(housekeeping gene), which measures the differential expression. This is to essentially normalise the gene of interest to a gene which is not affected by the experiment.</p>
</div>
<div id="questions-to-be-addressed" class="section level1">
<h1>Questions to be addressed</h1>
<ol style="list-style-type: decimal">
<li>Do gene expressions differ among genotypes? treatment? interactions between genetypes and treatment?</li>
</ol>
<p>– Run a two-way anova with interaction and adjustment for unbalance.</p>
<ol start="2" style="list-style-type: decimal">
<li>If they differ, how?</li>
</ol>
<p>– run tukey’s multiple comparisons.</p>
</div>
<div id="method" class="section level1">
<h1>Method</h1>
<div id="delta-delta-ct-method" class="section level2">
<h2>Delta-delta Ct method</h2>
<p>Delta-delta method is a very popular approach for qPCR data analysis which tells you the fold change of the gene expression. The output is <span class="math inline">\(2^{-\Delta\Delta Ct}\)</span>, where <span class="math inline">\(\Delta\Delta Ct\)</span> is <span class="math inline">\(\Delta Ct(treatment)-\Delta Ct(control\hspace{1mm}average)\)</span>. The fold change is the expression ratio: if the fold change is positive it means that the gene is upregulated; if the fold change is negative it means it is downregulated.</p>
</div>
<div id="directly-use-delta-ct" class="section level2">
<h2>Directly use Delta Ct</h2>
<p>I sugget directly use Delte Ct values instead of going for <span class="math inline">\(2^{-\Delta\Delta Ct}\)</span>. The first reason is that to run statistical hypothesis on <span class="math inline">\(2^{-\Delta\Delta Ct}\)</span>, we need to take <span class="math inline">\(\log\)</span> trasformation on it, which is equivalent to use <span class="math inline">\(-\Delta\Delta Ct\)</span>. Further, runing anova and other statistical tests on <span class="math inline">\(-\Delta\Delta Ct\)</span> is equivalent to runing them on <span class="math inline">\(\Delta Ct\)</span>, since <span class="math inline">\(-\Delta\Delta Ct\)</span> is obtained just by subtrating a number from it. By saying equivalent, it’s in terms of the statistical significance. Second, Delta-delta Ct method is mainly for calculating fold change of gene expressions see if gene is upregulated or not. We can obtain the same results from running statisitcal tests on <span class="math inline">\(\Delta Ct\)</span>.</p>
</div>
</div>
<div id="analysis" class="section level1">
<h1>Analysis</h1>
<div id="igha" class="section level2">
<h2>Igha</h2>
<pre class="r"><code>library(car)</code></pre>
<pre><code>Loading required package: carData</code></pre>
<pre class="r"><code>datax = read.csv(&#39;~/Downloads/Igha.csv&#39;,header = TRUE,skip=1)

# Use Treatment, Genotype and Normzalized value

datax = datax[,c(7,3,4)]

datax$Treatment = factor(datax$Treatment,levels = c(&#39;SPF&#39;,&#39;CMA&#39;, &#39;HEALTHY&#39;))

#table(datax$Genotype,datax$Treatment)

# fit 2-way anova
anova.fit = aov(dCT ~ Treatment * Genotype, data = datax)

# check normal asusmptions

#plot(anova.fit,1)
#plot(anova.fit,2)</code></pre>
<p>Variance is homogeneous and dCT is normal distributed. So we do not need to do transformations of the dCT.</p>
<pre class="r"><code># adjust due to unbanlanced design
Anova(anova.fit,type=&#39;III&#39;)</code></pre>
<pre><code>Anova Table (Type III tests)

Response: dCT
                    Sum Sq Df F value    Pr(&gt;F)    
(Intercept)        242.739  1 56.8959 1.368e-08 ***
Treatment           24.131  2  2.8280   0.07397 .  
Genotype             7.399  1  1.7344   0.19721    
Treatment:Genotype   1.250  2  0.1465   0.86432    
Residuals          136.524 32                      
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
<p>This suggests that Igha gene expression does not differ significantly among genotypes and treatments.</p>
</div>
<div id="reg3b" class="section level2">
<h2>Reg3b</h2>
<pre class="r"><code>datax = read.csv(&#39;~/Downloads/Reg3b.csv&#39;,header = TRUE,skip=1)

# Use Treatment, Genotype and Normzalized value

datax = datax[,c(7,3,4)]
datax$Treatment = factor(datax$Treatment,levels = c(&#39;SPF&#39;,&#39;CMA&#39;, &#39;HEALTHY&#39;))

#table(datax$Genotype,datax$Treatment)

# fit 2-way anova
anova.fit = aov(dCT ~ Treatment * Genotype, data = datax)

# check asusmptions

#plot(anova.fit,1)
#plot(anova.fit,2)

Anova(anova.fit,type=&#39;III&#39;)</code></pre>
<pre><code>Anova Table (Type III tests)

Response: dCT
                   Sum Sq Df  F value    Pr(&gt;F)    
(Intercept)        320.67  1 219.8846 6.868e-16 ***
Treatment           31.90  2  10.9361 0.0002402 ***
Genotype             1.34  1   0.9205 0.3445361    
Treatment:Genotype   7.35  2   2.5184 0.0964448 .  
Residuals           46.67 32                       
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
<p>This suggests: Treatment is significantly associated with Reg3b gene expression.</p>
<p>Then we answer the question: how do gene expressions differ among treatment. We run tukey’s multiple comparison tests on treatment:</p>
<pre class="r"><code>TukeyHSD(anova.fit, which = &quot;Treatment&quot;)</code></pre>
<pre><code>  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = dCT ~ Treatment * Genotype, data = datax)

$Treatment
                  diff        lwr        upr     p adj
CMA-SPF     -1.9341713 -3.1298418 -0.7385008 0.0010664
HEALTHY-SPF -0.4696847 -1.6854212  0.7460519 0.6135921
HEALTHY-CMA  1.4644866  0.3214818  2.6074914 0.0096486</code></pre>
<p>From the adjusted p-value(p adj), we conlude that gene expressions differ significantly between SPF and CMA as well as HEALTHY and CMA, but not between SPF and HEALTHY. The column diff gives the difference between means of the two treatments.</p>
</div>
<div id="jchain" class="section level2">
<h2>Jchain</h2>
<pre class="r"><code>datax = read.csv(&#39;~/Downloads/Jchain.csv&#39;,header = TRUE,skip=1)

# Use Treatment, Genotype and Normzalized value

datax = datax[,c(7,3,4)]
datax$Treatment = factor(datax$Treatment,levels = c(&#39;SPF&#39;,&#39;CMA&#39;, &#39;HEALTHY&#39;))

#table(datax$Genotype,datax$Treatment)

# fit 2-way anova
anova.fit = aov(dCT ~ Treatment * Genotype, data = datax)

# check asusmptions

#plot(anova.fit,1)
#plot(anova.fit,2)

Anova(anova.fit,type=&#39;III&#39;)</code></pre>
<pre><code>Anova Table (Type III tests)

Response: dCT
                    Sum Sq Df  F value    Pr(&gt;F)    
(Intercept)        225.065  1 100.9293 2.024e-11 ***
Treatment           19.876  2   4.4566   0.01962 *  
Genotype             0.158  1   0.0708   0.79189    
Treatment:Genotype   3.687  2   0.8266   0.44666    
Residuals           71.358 32                       
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
<p>This suggests Treatment is significantly associated with Jchain gene expression.</p>
<pre class="r"><code>TukeyHSD(anova.fit, which = &quot;Treatment&quot;)</code></pre>
<pre><code>  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = dCT ~ Treatment * Genotype, data = datax)

$Treatment
                  diff        lwr       upr     p adj
CMA-SPF     -1.3075130 -2.7860283 0.1710024 0.0913477
HEALTHY-SPF  0.2387343 -1.2645939 1.7420624 0.9196956
HEALTHY-CMA  1.5462473  0.1328561 2.9596384 0.0296045</code></pre>
<p>From the adjusted p-value(p adj), we conlude that gene expressions differ significantly between SPF and CMA as well as HEALTHY and CMA, but not between SPF and HEALTHY. The column diff gives the difference between means of the two treatments.</p>
</div>
<div id="saa1" class="section level2">
<h2>Saa1</h2>
<pre class="r"><code>datax = read.csv(&#39;~/Downloads/Saa1.csv&#39;,header = TRUE,skip=1)

# Use Treatment, Genotype and Normzalized value

datax = datax[,c(7,3,4)]
datax$Treatment = factor(datax$Treatment,levels = c(&#39;SPF&#39;,&#39;CMA&#39;, &#39;HEALTHY&#39;))
#table(datax$Genotype,datax$Treatment)

# fit 2-way anova
anova.fit = aov(dCT ~ Treatment * Genotype, data = datax)

# check asusmptions
#plot(anova.fit,1)
#plot(anova.fit,2)

Anova(anova.fit,type=&#39;III&#39;)</code></pre>
<pre><code>Anova Table (Type III tests)

Response: dCT
                    Sum Sq Df F value   Pr(&gt;F)    
(Intercept)        169.905  1 48.0357 7.56e-08 ***
Treatment          118.261  2 16.7174 1.07e-05 ***
Genotype             1.112  1  0.3143  0.57895    
Treatment:Genotype  27.661  2  3.9102  0.03025 *  
Residuals          113.186 32                     
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
<p>This suggests that Treatment is significantly associated with Saa1 gene expression. Also, the relationship between Saa1 gene expression and treatment depends on genotype.</p>
<pre class="r"><code>TukeyHSD(anova.fit,which = &#39;Treatment&#39;)</code></pre>
<pre><code>  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = dCT ~ Treatment * Genotype, data = datax)

$Treatment
                 diff         lwr        upr     p adj
CMA-SPF     -4.535648 -6.39773996 -2.6735570 0.0000033
HEALTHY-SPF -2.838310 -4.73165177 -0.9449686 0.0023680
HEALTHY-CMA  1.697338 -0.08273368  3.4774103 0.0640796</code></pre>
<p>From the adjusted p-value(p adj), we conlude that gene expressions differ significantly between SPF and CMA as well as HEALTHY and SPF, but not between CMA and HEALTHY. The column diff gives the difference between means of the two treatments.</p>
<p>To visulize the interaction, we can plot</p>
<pre class="r"><code>interaction.plot(x.factor = datax$Treatment, trace.factor = datax$Genotype, 
                 response = datax$dCT, fun = mean, 
                 type = &quot;b&quot;, legend = TRUE, 
                 xlab = &quot;Treatment&quot;, ylab=&quot;dCt&quot;,
                 pch=c(1,19), col = c(&quot;#00AFBB&quot;, &quot;#E7B800&quot;))</code></pre>
<p><img src="figure/consulting2.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>To have a detailed look of all interations:</p>
<pre class="r"><code>TukeyHSD(anova.fit,which=&#39;Treatment:Genotype&#39;)</code></pre>
<pre><code>  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = dCT ~ Treatment * Genotype, data = datax)

$`Treatment:Genotype`
                                                  diff        lwr
CMA:CD11c-cre-neg-SPF:CD11c-cre-neg         -5.4955206 -8.4439766
HEALTHY:CD11c-cre-neg-SPF:CD11c-cre-neg     -1.7969384 -4.9664333
SPF:CD11c-cre-pos-SPF:CD11c-cre-neg         -0.6608739 -4.2316314
CMA:CD11c-cre-pos-SPF:CD11c-cre-neg         -3.8165604 -6.9860553
HEALTHY:CD11c-cre-pos-SPF:CD11c-cre-neg     -4.1772190 -7.2223712
HEALTHY:CD11c-cre-neg-CMA:CD11c-cre-neg      3.6985822  0.6218732
SPF:CD11c-cre-pos-CMA:CD11c-cre-neg          4.8346468  1.3459866
CMA:CD11c-cre-pos-CMA:CD11c-cre-neg          1.6789603 -1.3977487
HEALTHY:CD11c-cre-pos-CMA:CD11c-cre-neg      1.3183017 -1.6301543
SPF:CD11c-cre-pos-HEALTHY:CD11c-cre-neg      1.1360645 -2.5413061
CMA:CD11c-cre-pos-HEALTHY:CD11c-cre-neg     -2.0196219 -5.3087622
HEALTHY:CD11c-cre-pos-HEALTHY:CD11c-cre-neg -2.3802805 -5.5497754
CMA:CD11c-cre-pos-SPF:CD11c-cre-pos         -3.1556865 -6.8330571
HEALTHY:CD11c-cre-pos-SPF:CD11c-cre-pos     -3.5163451 -7.0871025
HEALTHY:CD11c-cre-pos-CMA:CD11c-cre-pos     -0.3606586 -3.5301535
                                                   upr     p adj
CMA:CD11c-cre-neg-SPF:CD11c-cre-neg         -2.5470647 0.0000420
HEALTHY:CD11c-cre-neg-SPF:CD11c-cre-neg      1.3725565 0.5308744
SPF:CD11c-cre-pos-SPF:CD11c-cre-neg          2.9098836 0.9928809
CMA:CD11c-cre-pos-SPF:CD11c-cre-neg         -0.6470655 0.0109812
HEALTHY:CD11c-cre-pos-SPF:CD11c-cre-neg     -1.1320668 0.0028441
HEALTHY:CD11c-cre-neg-CMA:CD11c-cre-neg      6.7752913 0.0111569
SPF:CD11c-cre-pos-CMA:CD11c-cre-neg          8.3233069 0.0025316
CMA:CD11c-cre-pos-CMA:CD11c-cre-neg          4.7556693 0.5711781
HEALTHY:CD11c-cre-pos-CMA:CD11c-cre-neg      4.2667576 0.7528749
SPF:CD11c-cre-pos-HEALTHY:CD11c-cre-neg      4.8134352 0.9340990
CMA:CD11c-cre-pos-HEALTHY:CD11c-cre-neg      1.2695184 0.4439477
HEALTHY:CD11c-cre-pos-HEALTHY:CD11c-cre-neg  0.7892144 0.2336856
CMA:CD11c-cre-pos-SPF:CD11c-cre-pos          0.5216842 0.1267106
HEALTHY:CD11c-cre-pos-SPF:CD11c-cre-pos      0.0544124 0.0555629
HEALTHY:CD11c-cre-pos-CMA:CD11c-cre-pos      2.8088363 0.9992932</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.1 (2019-07-05)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] car_3.0-5     carData_3.0-3

loaded via a namespace (and not attached):
 [1] zip_2.0.4         Rcpp_1.0.2        cellranger_1.1.0 
 [4] compiler_3.6.1    pillar_1.4.2      later_1.0.0      
 [7] git2r_0.26.1      workflowr_1.5.0   forcats_0.4.0    
[10] tools_3.6.1       zeallot_0.1.0     digest_0.6.21    
[13] evaluate_0.14     tibble_2.1.3      pkgconfig_2.0.3  
[16] rlang_0.4.0       openxlsx_4.1.4    curl_4.2         
[19] yaml_2.2.0        haven_2.2.0       xfun_0.10        
[22] rio_0.5.16        stringr_1.4.0     knitr_1.25       
[25] fs_1.3.1          vctrs_0.2.0       hms_0.5.2        
[28] rprojroot_1.3-2   glue_1.3.1        data.table_1.12.6
[31] R6_2.4.0          readxl_1.3.1      foreign_0.8-72   
[34] rmarkdown_1.16    magrittr_1.5      whisker_0.4      
[37] backports_1.1.5   promises_1.1.0    htmltools_0.4.0  
[40] abind_1.4-5       httpuv_1.5.2      stringi_1.4.3    
[43] crayon_1.3.4     </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
