<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Dongyue Xie" />

<meta name="date" content="2020-01-04" />

<title>Common transformation of counts</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.0.13/css/fa-svg-with-js.css" rel="stylesheet" />
<script src="site_libs/font-awesome-5.0.13/js/fontawesome-all.min.js"></script>
<script src="site_libs/font-awesome-5.0.13/js/fa-v4-shims.min.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">misc</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/DongyueXie/misc">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Common transformation of counts</h1>
<h4 class="author"><em>Dongyue Xie</em></h4>
<h4 class="date"><em>2020-01-04</em></h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-09-21
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 2 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>misc/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomDongyueXiemisctree93f539537c5ae0a57dd0e7edb71080415f585255targetblank93f5395a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/DongyueXie/misc/tree/93f539537c5ae0a57dd0e7edb71080415f585255" target="_blank">93f5395</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomDongyueXiemisctree93f539537c5ae0a57dd0e7edb71080415f585255targetblank93f5395a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/DongyueXie/misc/tree/93f539537c5ae0a57dd0e7edb71080415f585255" target="_blank">93f5395</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  analysis/contrainedclustering.Rmd
    Untracked:  analysis/count.Rmd
    Untracked:  analysis/deconvSimulation2.Rmd
    Untracked:  analysis/ideas.Rmd
    Untracked:  analysis/methylation.Rmd
    Untracked:  analysis/ridge_em.Rmd
    Untracked:  code/sccytokines.R
    Untracked:  code/scdeCalibration.R
    Untracked:  data/bart/
    Untracked:  data/cytokine/DE_controls_output_filter10.RData
    Untracked:  data/cytokine/DE_controls_output_filter10_addlimma.RData
    Untracked:  data/cytokine/README
    Untracked:  data/cytokine/test.RData
    Untracked:  data/cytokine_normalized.RData
    Untracked:  data/deconv/
    Untracked:  data/scde/

Unstaged changes:
    Modified:   analysis/deconvSimulation.Rmd
    Modified:   analysis/deconvolution.Rmd
    Deleted:    data/mout_high.RData
    Deleted:    data/scCDT.RData
    Deleted:    data/sva_sva_high.RData

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/CPM.Rmd</code>) and HTML (<code>docs/CPM.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DongyueXie/misc/blob/93f539537c5ae0a57dd0e7edb71080415f585255/analysis/CPM.Rmd" target="_blank">93f5395</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-09-21
</td>
<td>
wflow_publish(c(“analysis/CPM.Rmd”, “analysis/limma.Rmd”))
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DongyueXie/misc/6bb0476b6ea5565949270fbf7477e2b33f500b3e/docs/CPM.html" target="_blank">6bb0476</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-06-09
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DongyueXie/misc/blob/0143800833bb599f5e2e90f8f26fe0872c0e424e/analysis/CPM.Rmd" target="_blank">0143800</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-06-09
</td>
<td>
wflow_publish(“analysis/CPM.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DongyueXie/misc/c89c7f9cf0f1126ee7cde0c1155e4a8312384414/docs/CPM.html" target="_blank">c89c7f9</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-13
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DongyueXie/misc/blob/f4f3ac6a60d389073acab6951f984308faf9d38e/analysis/CPM.Rmd" target="_blank">f4f3ac6</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-13
</td>
<td>
wflow_publish(“analysis/CPM.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DongyueXie/misc/d656bc907f6e9c079ae48f877a0836a8b99f89a3/docs/CPM.html" target="_blank">d656bc9</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-13
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DongyueXie/misc/blob/47f7d2a78b79a1bfebb675c896bc506826b32f5e/analysis/CPM.Rmd" target="_blank">47f7d2a</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-13
</td>
<td>
wflow_publish(“analysis/CPM.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DongyueXie/misc/2a4e9aa72a9d1af343b46bb5f913363e04945086/docs/CPM.html" target="_blank">2a4e9aa</a>
</td>
<td>
Dongyue Xie
</td>
<td>
2020-01-06
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DongyueXie/misc/0452cba01c51844572c6f57e8c10e8bd1b8072b3/docs/CPM.html" target="_blank">0452cba</a>
</td>
<td>
Dongyue Xie
</td>
<td>
2020-01-04
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DongyueXie/misc/blob/e033ca37a64a7ff54aefb3a855a3813c115cd8e9/analysis/CPM.Rmd" target="_blank">e033ca3</a>
</td>
<td>
Dongyue Xie
</td>
<td>
2020-01-04
</td>
<td>
wflow_publish(“analysis/CPM.Rmd”)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p><a href="https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/">A good review</a> of RNA-Seq expression units from <a href="https://pimentel.github.io/">Harold Pimentel</a>.</p>
<p>CPM: counts per million</p>
<p>RPKM: reads per kilobase per million</p>
<p>In general, need to consider 1. library size; 2. gene length. Because libraries sequenced at a greater depth will result in higher counts.</p>
<p>CPM and log-CPM transformations do not account for gene length differences as RPKM and FPKM values do. CPM and log-CPM values can be calculated using a counts matrix alone. Assuming that there are no differences in isoform usage between conditions, differential expression analyses look at gene expression changes between conditions rather than comparing expression across multiple genes or drawing conclusions on absolute levels of expression. In other words, gene lengths remain constant for comparisons of interest and any observed differences are a result of changes in condition rather than changes in gene length.</p>
<p>RPKM values are just as easily calculated as CPM values using the rpkm function in edgeR if gene lengths are available.</p>
</div>
<div id="trimmed-mean-of-m-values" class="section level2">
<h2>Trimmed mean of M values</h2>
<p>Observed counts as <span class="math inline">\(Y_{gk}\)</span>, gene length <span class="math inline">\(l_g\)</span>, true and unknown expression level (number of transcripts) <span class="math inline">\(\mu_{gk}\)</span>, total number of reads for library k <span class="math inline">\(N_k\)</span>, let <span class="math inline">\(s_k = \sum_g l_g \mu_{gk}\)</span>, then <span class="math display">\[E(Y_{gk}) = \frac{\mu_{gk}l_g}{s_k}N_k\]</span></p>
<p>The ratio of expected value between two conditions is <span class="math display">\[\frac{E(Y_{g1})}{E(Y_{g2})} = \frac{s_1}{s_2}\frac{N_2}{N_1}\frac{\mu_{g2}}{\mu_{g1}}\]</span></p>
<p>So <span class="math inline">\(s_k\)</span> is the size of studied transcriptome in condition <span class="math inline">\(k\)</span>, <span class="math inline">\(\mu_{gk}/s_k\)</span> is the relative expression of gene <span class="math inline">\(g\)</span> in condition <span class="math inline">\(k\)</span>.</p>
<p>Let <span class="math inline">\(r\)</span> denote the reference sample. Filter out 1. transcripts with null counts(The cases where <span class="math inline">\(Y_{gk} = 0\)</span> or <span class="math inline">\(Y_{gr} = 0\)</span> are trimmed); 2. the <span class="math inline">\(30\%\)</span> more extreme <span class="math inline">\(M-value\)</span>: <span class="math inline">\(M_{gk}^r = \log\frac{y_{gk}/N_k}{y_{gr}/N_r}\)</span>; 3. the <span class="math inline">\(5\%\)</span> more extreme <span class="math inline">\(A-value\)</span>: <span class="math inline">\(A_{gk}^r = 0.5*(\log\frac{y_{gk}}{N_k}+\log\frac{y_{gr}}{N_r})\)</span>.</p>
<p>After filtering out genes, we have a set of genes denoted as <span class="math inline">\(G^*\)</span> whose neither value was trimmed. The scaling factor <span class="math inline">\(TMM_k\)</span> is calculated as <span class="math display">\[\log TMM_k = \frac{\sum_{g\in G^*}w_{gk}M_{gk}}{\sum_{g\in G^*}w_{gk}},\]</span> where <span class="math inline">\(w_{gk} = \frac{N_k-Y_{gk}}{N_k Y_{gk}} - \frac{N_r-Y_{gr}}{N_r Y_{gr}}\)</span>. Usually the scaling factors are re-scaled such that they multiply to 1.</p>
</div>
<div id="relative-log-expression" class="section level2">
<h2>Relative Log Expression</h2>
<p>Define the reference as the geometric mean(less sensitive to extreme values than the standard mean) of samples <span class="math inline">\(Y_{gk}^r = (\Pi_k Y_{gk})^{1/n}\)</span>, then the scaling factor for sample <span class="math inline">\(k\)</span> is calculated as <span class="math inline">\(median(\frac{Y_{gk}}{Y_{gk}^r})\)</span>.</p>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
