<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Dongyue Xie" />

<meta name="date" content="2019-12-10" />

<title>Deconvolution</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.0.13/css/fa-svg-with-js.css" rel="stylesheet" />
<script src="site_libs/font-awesome-5.0.13/js/fontawesome-all.min.js"></script>
<script src="site_libs/font-awesome-5.0.13/js/fa-v4-shims.min.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">misc</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/DongyueXie/misc">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Deconvolution</h1>
<h4 class="author"><em>Dongyue Xie</em></h4>
<h4 class="date"><em>2019-12-10</em></h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-04-02
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>misc/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20191122code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20191122)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20191122code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20191122)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongabsolute"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>File paths:</strong> absolute </a>
</p>
</div>
<div id="strongFilepathsstrongabsolute" class="panel-collapse collapse">
<div class="panel-body">
<p>
Using absolute paths to the files within your workflowr project makes it difficult for you and others to run your code on a different machine. Change the absolute path(s) below to the suggested relative path(s) to make your code more reproducible.
</p>
<table class="table table-condensed table-hover">
<thead>
<tr>
<th style="text-align:left;">
absolute
</th>
<th style="text-align:left;">
relative
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
~/misc/data/deconv/GSE50244bulkeset.rds
</td>
<td style="text-align:left;">
data/deconv/GSE50244bulkeset.rds
</td>
</tr>
<tr>
<td style="text-align:left;">
~/misc/data/deconv/Mousebulkeset.rds
</td>
<td style="text-align:left;">
data/deconv/Mousebulkeset.rds
</td>
</tr>
<tr>
<td style="text-align:left;">
~/misc/data/scde/scCD14.RData
</td>
<td style="text-align:left;">
data/scde/scCD14.RData
</td>
</tr>
<tr>
<td style="text-align:left;">
~/misc/data/scde/scCD4.RData
</td>
<td style="text-align:left;">
data/scde/scCD4.RData
</td>
</tr>
<tr>
<td style="text-align:left;">
~/misc/data/scde/scCD8.RData
</td>
<td style="text-align:left;">
data/scde/scCD8.RData
</td>
</tr>
<tr>
<td style="text-align:left;">
~/misc/data/scde/scCDT.RData
</td>
<td style="text-align:left;">
data/scde/scCDT.RData
</td>
</tr>
<tr>
<td style="text-align:left;">
~/misc/data/scde/scMB.RData
</td>
<td style="text-align:left;">
data/scde/scMB.RData
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomDongyueXiemisctree78f2fdc0741a6ac01fbc65e939740c4bf7d63f22targetblank78f2fdca"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/DongyueXie/misc/tree/78f2fdc0741a6ac01fbc65e939740c4bf7d63f22" target="_blank">78f2fdc</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomDongyueXiemisctree78f2fdc0741a6ac01fbc65e939740c4bf7d63f22targetblank78f2fdca" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  analysis/contrainedclustering.Rmd
    Untracked:  analysis/gsea.Rmd
    Untracked:  analysis/ideas.Rmd
    Untracked:  analysis/methylation.Rmd
    Untracked:  analysis/susie.Rmd
    Untracked:  code/sccytokines.R
    Untracked:  code/scdeCalibration.R
    Untracked:  data/bart/
    Untracked:  data/cytokine/DE_controls_output_filter10.RData
    Untracked:  data/cytokine/DE_controls_output_filter10_addlimma.RData
    Untracked:  data/cytokine/README
    Untracked:  data/cytokine/test.RData
    Untracked:  data/cytokine_normalized.RData
    Untracked:  data/deconv/
    Untracked:  data/scde/

Unstaged changes:
    Modified:   analysis/CPM.Rmd
    Modified:   analysis/index.Rmd
    Modified:   analysis/limma.Rmd
    Deleted:    data/mout_high.RData
    Deleted:    data/scCDT.RData
    Deleted:    data/sva_sva_high.RData

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DongyueXie/misc/blob/78f2fdc0741a6ac01fbc65e939740c4bf7d63f22/analysis/deconvolution.Rmd" target="_blank">78f2fdc</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-04-02
</td>
<td>
wflow_publish(“analysis/deconvolution.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DongyueXie/misc/90dde67cda8041a24a5548fa8154ddbacf8ade23/docs/deconvolution.html" target="_blank">90dde67</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-03-26
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DongyueXie/misc/blob/2ab0aa9e618f9db36c18228ebb28d53d062fc771/analysis/deconvolution.Rmd" target="_blank">2ab0aa9</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-03-26
</td>
<td>
wflow_publish(“analysis/deconvolution.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DongyueXie/misc/a617635adcc9a21437ce6c50f2b57a2927fd43f9/docs/deconvolution.html" target="_blank">a617635</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-03-13
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DongyueXie/misc/blob/73933a7941216795b4a58fc8f6431710a573e281/analysis/deconvolution.Rmd" target="_blank">73933a7</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-03-13
</td>
<td>
wflow_publish(“analysis/deconvolution.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DongyueXie/misc/53cef0f15bd29ba0cd6cdcd8060868b56b5cf0bd/docs/deconvolution.html" target="_blank">53cef0f</a>
</td>
<td>
Dongyue Xie
</td>
<td>
2019-12-10
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DongyueXie/misc/blob/af20c492be164276a63bcf2f1330b0bf99d3ec8b/analysis/deconvolution.Rmd" target="_blank">af20c49</a>
</td>
<td>
Dongyue Xie
</td>
<td>
2019-12-10
</td>
<td>
wflow_publish(“analysis/deconvolution.Rmd”)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>check bulk data library size: vary from few hundred(100-200) to few thousands(1600). Here</p>
<pre class="r"><code>GSE50244bulkeset &lt;- readRDS(&quot;~/misc/data/deconv/GSE50244bulkeset.rds&quot;)
GSE50244bulkeset</code></pre>
<pre><code>ExpressionSet (storageMode: lockedEnvironment)
assayData: 32581 features, 89 samples 
  element names: exprs 
protocolData: none
phenoData
  sampleNames: Sub1 Sub2 ... Sub89 (89 total)
  varLabels: sampleID SubjectName ... tissue (7 total)
  varMetadata: labelDescription
featureData: none
experimentData: use &#39;experimentData(object)&#39;
Annotation:  </code></pre>
<pre class="r"><code>summary(as.numeric(apply(exprs(GSE50244bulkeset),2,mean)))</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  376.3   506.4   605.8   587.1   664.4   793.6 </code></pre>
<pre class="r"><code>Mousebulkeset &lt;- readRDS(&quot;~/misc/data/deconv/Mousebulkeset.rds&quot;)
summary(as.numeric(apply(exprs(Mousebulkeset),2,mean)))</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   1431    1546    1608    1625    1729    1805 </code></pre>
<p>We have scRNA-seq data from Segerstolpe et al. (2016) including the 1097 cells from 6 healthy subjects, taken from <a href="https://xuranw.github.io/MuSiC/articles/pages/data.html">here</a>. This dataset is non-UMI so the read counts are huge. In the following simulation study, 4 cell types - acinar, alpha, beta, and ductal cells are included.</p>
<pre class="r"><code>EMTAB.eset = readRDS(&#39;data/deconv/EMTABesethealthy.rds&#39;)
EMTAB.eset</code></pre>
<pre><code>ExpressionSet (storageMode: lockedEnvironment)
assayData: 25453 features, 1097 samples 
  element names: exprs 
protocolData: none
phenoData
  sampleNames: AZ_A10 AZ_A11 ... HP1509101_P9 (1097 total)
  varLabels: sampleID SubjectName cellTypeID cellType
  varMetadata: labelDescription
featureData: none
experimentData: use &#39;experimentData(object)&#39;
Annotation:  </code></pre>
<pre class="r"><code>table(EMTAB.eset$cellType)</code></pre>
<pre><code>
                acinar                  alpha                   beta 
                   112                    443                    171 
         co-expression                  delta                 ductal 
                    26                     59                    135 
           endothelial                epsilon                  gamma 
                    13                      5                     75 
                  mast           MHC class II                    PSC 
                     4                      1                     23 
          unclassified unclassified endocrine 
                     1                     29 </code></pre>
<p>Now obtain the true gene relative expression in each cell type.</p>
<p>Preprocess the data: 1. remove genes appearing in less than 10 cells; 2. remove top <span class="math inline">\(5\%\)</span> expressed genes</p>
<pre class="r"><code>cell_types = c(&#39;acinar&#39;,&#39;alpha&#39;,&#39;beta&#39;,&#39;ductal&#39;)

###remove genes appeared in too few cells
###remove genes that are overly expressed
#rm.idx = which(rowSums((exprs(EMTAB.eset)[,which(EMTAB.eset$cellType%in%cell_types)])!=0)&lt;10)
#rr = rowSums((exprs(EMTAB.eset)[,which(EMTAB.eset$cellType%in%cell_types)]))
###rm.idx = which(rr==0)
###rm.idx1 = which(rr&lt;=quantile(rr[-rm.idx],0.01))
#rm.idx2 = which(rr&gt;=quantile(rr[-rm.idx],0.99))
#rm.idx = unique(c(rm.idx,rm.idx2))

acinar = exprs(EMTAB.eset)[,which(EMTAB.eset$cellType==&#39;acinar&#39;)]
alpha = exprs(EMTAB.eset)[,which(EMTAB.eset$cellType==&#39;alpha&#39;)]
beta = exprs(EMTAB.eset)[,which(EMTAB.eset$cellType==&#39;beta&#39;)]
ductal = exprs(EMTAB.eset)[,which(EMTAB.eset$cellType==&#39;ductal&#39;)]

rm.idx1 = which(((rowSums(acinar!=0)&lt;=10)|(rowSums(acinar)&gt;=quantile(rowSums(acinar),0.95))))
rm.idx2 = which(((rowSums(alpha!=0)&lt;=10)|(rowSums(alpha)&gt;=quantile(rowSums(alpha),0.95))))
rm.idx3 = which(((rowSums(beta!=0)&lt;=10)|(rowSums(beta)&gt;=quantile(rowSums(beta),0.95))))
rm.idx4 = which(((rowSums(ductal!=0)&lt;=10)|(rowSums(ductal)&gt;=quantile(rowSums(ductal),0.95))))

rm.idx = unique(c(rm.idx1,rm.idx2,rm.idx3,rm.idx4))

acinar = acinar[-rm.idx,]
alpha = alpha[-rm.idx,]
beta = beta[-rm.idx,]
ductal = ductal[-rm.idx,]</code></pre>
<p>Check cell library size: cell library sizes are big</p>
<pre class="r"><code>summary(colSums(acinar))</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   2686   40757   75941  129023  186640  553802 </code></pre>
<pre class="r"><code>summary(colSums(alpha))</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   3523   40465   83358  113020  159609  690603 </code></pre>
<pre class="r"><code>summary(colSums(beta))</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   2970   47179  100240  124888  190024  537763 </code></pre>
<pre class="r"><code>summary(colSums(ductal))</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   6405   38504   86323  135121  191832  739986 </code></pre>
<pre class="r"><code># cell type specific gene relative expression
Theta = cbind(rowSums(acinar)/sum(acinar),
              rowSums(alpha)/sum(alpha),
              rowSums(beta)/sum(beta),
              rowSums(ductal)/sum(ductal))

cor(Theta)</code></pre>
<pre><code>          [,1]      [,2]      [,3]      [,4]
[1,] 1.0000000 0.6653738 0.5972479 0.8288408
[2,] 0.6653738 1.0000000 0.8156478 0.6819325
[3,] 0.5972479 0.8156478 1.0000000 0.6363127
[4,] 0.8288408 0.6819325 0.6363127 1.0000000</code></pre>
<pre class="r"><code>kappa(t(Theta)%*%Theta)</code></pre>
<pre><code>[1] 34.74529</code></pre>
<pre class="r"><code># cell library size
S = c(sum(acinar)/ncol(acinar),
      sum(alpha)/ncol(alpha),
      sum(beta)/ncol(beta),
      sum(ductal)/ncol(ductal))

#S=S/100 </code></pre>
<p>Generate reference data: 1. obtain <span class="math inline">\(\tilde{\theta}_{gk}\)</span> by summing up read counts and normalize; 2. obtain <span class="math inline">\(y_{k}^{r,+}\)</span> by multiple cell library size and number of cells(set to 100); 3. Generate reference data using the Poisson model</p>
<p>Generate bulk data: 1. simulate library size <span class="math inline">\(y^{b,+}\)</span> from Poisson<span class="math inline">\((300G)\)</span>; 2. generate <span class="math inline">\(x^b_g\)</span> by multiplying <span class="math inline">\(N=400\)</span>, <span class="math inline">\(\beta = (0.1,0.2,0.3,0.4)\)</span>, and <span class="math inline">\(S\)</span> cell library size then normalize to <span class="math inline">\(\theta^b_g\)</span>. 3. generate bulk data using Poisson distribution.</p>
<pre class="r"><code>G = nrow(Theta)
K = ncol(Theta)

set.seed(12345)
# bulk data library size: 300*number of genes.
# bulk_ls = rpois(1,300*nrow(Theta))
# total number of cells in bulk data
bulk_ncell = 400
# cell proportions
bulk_beta = c(1,2,3,4)
bulk_beta = bulk_beta/sum(bulk_beta)
bulk_p = (bulk_beta * (S))/sum(bulk_beta * (S))
# bulk data gene relative expression.
bulk_X = bulk_ncell*Theta%*%diag(S)%*%bulk_beta
bulk_theta = bulk_X/sum(bulk_X)

ref_ncell = 100
ref_X = Theta%*%diag(S)*ref_ncell

w = rep(1,G)


p_tilde_est = c()
p_tilde_var = c()
p_est = c()
p_var = c()

nrep = 100
for(rep in 1:nrep){
  if(rep%%200==0){print(rep)}
  bulk_ls = rpois(1,300*nrow(Theta))
  Z = matrix(rpois(G*K,ref_X),ncol=K)
  y = rpois(G,bulk_ls*bulk_theta)

  #W = diag(w)
  #beta_hat = solve(t(Z)%*%W%*%Z - diag(colSums(W%*%Z)))%*%t(Z)%*%W%*%y

  U = diag(colSums(Z))
  Zu = Z%*%(solve(U))
  Zuu = Zu%*%solve(U)
  p_tilde_hat = solve(t(Zu)%*%Zu - diag(colSums(Zuu)))%*%t(Zu)%*%y
  p_tilde_est = rbind(p_tilde_est,c(p_tilde_hat))
  #beta_hat = solve(t(Z)%*%Z)%*%t(Z)%*%y

  #beta_hat
  #beta_hat/sum(beta_hat)

  Q = 0
  Sigma=0
  deltas = c()
  for(i in 1:length(y)){ 
    ag = Zu[i,]%*%t(Zu[i,])-diag(Zuu[i,])
    Q = Q + ag
    Delta = (ag%*%p_tilde_hat-y[i]*Zu[i,])
    Sigma = Sigma + Delta%*%t(Delta)
  }
  Q = Q/G
  Sigma = Sigma/G

  J = matrix(nrow = K,ncol=K)
  for(i in 1:K){
    for(j in 1:K){
      if(i==j){
        J[i,j] = sum(p_tilde_hat)-p_tilde_hat[i]
      }else{
        J[i,j] = -p_tilde_hat[i]
      }
    }
  }

  J = J/sum(p_tilde_hat)^2

  asyV = ((J)%*%solve(Q)%*%Sigma%*%solve(Q)%*%t(J))/G

  p_hat = p_tilde_hat/sum(p_tilde_hat)
  p_est = rbind(p_est,c(p_hat))
  
  p_tilde_var = rbind(p_tilde_var,diag(solve(Q)%*%Sigma%*%solve(Q)))
  
  p_var = rbind(p_var,diag(asyV))

}


#### coverage before delta method:

#ci_l = p_tilde_est - qnorm(0.975)*sqrt(p_tilde_var)
#ci_r = p_tilde_est + qnorm(0.975)*sqrt(p_tilde_var)

#coverage

#coverage = ((rep(1,100)%*%t(bulk_p))&gt;=ci_l) &amp; ((rep(1,100)%*%t(bulk_p))&lt;=ci_r)
#apply(coverage,2,mean)

#### converage after delta method

ci_l = p_est - qnorm(0.975)*sqrt(p_var)
ci_r = p_est + qnorm(0.975)*sqrt(p_var)

#coverage

coverage = ((rep(1,100)%*%t(bulk_p))&gt;=ci_l) &amp; ((rep(1,100)%*%t(bulk_p))&lt;=ci_r)
apply(coverage,2,mean)</code></pre>
<pre><code>[1] 0.95 0.94 0.95 0.93</code></pre>
<pre class="r"><code>#for(i in 1:4){
#  plot(ci_l[,i],type=&#39;l&#39;,ylim=range(c(ci_l[,i],ci_r[,i])))
#  lines(ci_r[,i],col=&#39;grey80&#39;)
#  lines(rep(bulk_p[i],100),lty = 2)
#}</code></pre>
<p>how about decrease bulk library size to <span class="math inline">\(50G\)</span>?</p>
<pre class="r"><code>G = nrow(Theta)
K = ncol(Theta)

set.seed(12345)
# bulk data library size: 300*number of genes.
# bulk_ls = rpois(1,300*nrow(Theta))
# total number of cells in bulk data
bulk_ncell = 400
# cell proportions
bulk_beta = c(1,1,2,16)
bulk_beta = bulk_beta/sum(bulk_beta)
bulk_p = (bulk_beta * (S))/sum(bulk_beta * (S))
# bulk data gene relative expression.
bulk_X = bulk_ncell*Theta%*%diag(S)%*%bulk_beta
bulk_theta = bulk_X/sum(bulk_X)

ref_ncell = 100
ref_X = Theta%*%diag(S)*ref_ncell

w = rep(1,G)


p_tilde_est = c()
p_tilde_var = c()
p_est = c()
p_var = c()

nrep = 100
for(rep in 1:nrep){
  if(rep%%10==0){print(rep)}
  #bulk_ls = rpois(1,50*nrow(Theta))
  bulk_ls = 50*nrow(Theta)
  Z = matrix(rpois(G*K,ref_X),ncol=K)
  y = rpois(G,bulk_ls*bulk_theta)

  #W = diag(w)
  #beta_hat = solve(t(Z)%*%W%*%Z - diag(colSums(W%*%Z)))%*%t(Z)%*%W%*%y

  U = diag(colSums(Z))
  Zu = Z%*%(solve(U))
  Zuu = Zu%*%solve(U)
  p_tilde_hat = solve(t(Zu)%*%Zu - diag(colSums(Zuu)))%*%t(Zu)%*%y
  p_tilde_est = rbind(p_tilde_est,c(p_tilde_hat))
  #beta_hat = solve(t(Z)%*%Z)%*%t(Z)%*%y

  #beta_hat
  #beta_hat/sum(beta_hat)

  Q = 0
  Sigma=0
  deltas = c()
  for(i in 1:length(y)){ 
    ag = Zu[i,]%*%t(Zu[i,])-diag(Zuu[i,])
    Q = Q + ag
    Delta = (ag%*%p_tilde_hat-y[i]*Zu[i,])
    Sigma = Sigma + Delta%*%t(Delta)
  }
  Q = Q/G
  Sigma = Sigma/G

  J = matrix(nrow = K,ncol=K)
  for(i in 1:K){
    for(j in 1:K){
      if(i==j){
        J[i,j] = sum(p_tilde_hat)-p_tilde_hat[i]
        #J[i,j] = sum(p_tilde_hat)
      }else{
        J[i,j] = -p_tilde_hat[i]
        #J[i,j] = 0
      }
    }
  }

  J = J/sum(p_tilde_hat)^2

  asyV = ((J)%*%solve(Q)%*%Sigma%*%solve(Q)%*%t(J))/G

  p_hat = p_tilde_hat/sum(p_tilde_hat)
  p_est = rbind(p_est,c(p_hat))
  
  p_tilde_var = rbind(p_tilde_var,diag(solve(Q)%*%Sigma%*%solve(Q))/G)
  
  p_var = rbind(p_var,diag(asyV))

}</code></pre>
<pre><code>[1] 10
[1] 20
[1] 30
[1] 40
[1] 50
[1] 60
[1] 70
[1] 80
[1] 90
[1] 100</code></pre>
<pre class="r"><code>ci_l = p_est - qnorm(0.975)*sqrt(p_var)*4
ci_r = p_est + qnorm(0.975)*sqrt(p_var)*4

#coverage

coverage = ((rep(1,100)%*%t(bulk_p))&gt;=ci_l) &amp; ((rep(1,100)%*%t(bulk_p))&lt;=ci_r)
apply(coverage,2,mean)</code></pre>
<pre><code>[1] 1 1 1 1</code></pre>
<pre class="r"><code>#for(i in 1:4){
#  plot(ci_l[,i],type=&#39;l&#39;,ylim=range(c(ci_l[,i],ci_r[,i])))
#  lines(ci_r[,i],col=&#39;grey80&#39;)
#  lines(rep(bulk_p[i],100),lty = 2)
#}</code></pre>
<p>how about change <span class="math inline">\(beta\)</span> to be <span class="math inline">\((0.4,0.3,0.2,0.1)\)</span>?</p>
<pre class="r"><code>G = nrow(Theta)
K = ncol(Theta)

set.seed(12345)
# bulk data library size: 300*number of genes.
# bulk_ls = rpois(1,300*nrow(Theta))
# total number of cells in bulk data
bulk_ncell = 400
# cell proportions
bulk_beta = c(4,3,2,1)
bulk_beta = bulk_beta/sum(bulk_beta)
bulk_p = (bulk_beta * (S))/sum(bulk_beta * (S))
# bulk data gene relative expression.
bulk_X = bulk_ncell*Theta%*%diag(S)%*%bulk_beta
bulk_theta = bulk_X/sum(bulk_X)

ref_ncell = 100
ref_X = Theta%*%diag(S)*ref_ncell

w = rep(1,G)


p_tilde_est = c()
p_tilde_var = c()
p_est = c()
p_var = c()

nrep = 100
for(rep in 1:nrep){
  if(rep%%200==0){print(rep)}
  bulk_ls = rpois(1,300*nrow(Theta))
  Z = matrix(rpois(G*K,ref_X),ncol=K)
  y = rpois(G,bulk_ls*bulk_theta)

  #W = diag(w)
  #beta_hat = solve(t(Z)%*%W%*%Z - diag(colSums(W%*%Z)))%*%t(Z)%*%W%*%y

  U = diag(colSums(Z))
  Zu = Z%*%(solve(U))
  Zuu = Zu%*%solve(U)
  p_tilde_hat = solve(t(Zu)%*%Zu - diag(colSums(Zuu)))%*%t(Zu)%*%y
  p_tilde_est = rbind(p_tilde_est,c(p_tilde_hat))
  #beta_hat = solve(t(Z)%*%Z)%*%t(Z)%*%y

  #beta_hat
  #beta_hat/sum(beta_hat)

  Q = 0
  Sigma=0
  deltas = c()
  for(i in 1:length(y)){ 
    ag = Zu[i,]%*%t(Zu[i,])-diag(Zuu[i,])
    Q = Q + ag
    Delta = (ag%*%p_tilde_hat-y[i]*Zu[i,])
    Sigma = Sigma + Delta%*%t(Delta)
  }
  Q = Q/G
  Sigma = Sigma/G

  J = matrix(nrow = K,ncol=K)
  for(i in 1:K){
    for(j in 1:K){
      if(i==j){
        J[i,j] = sum(p_tilde_hat)-p_tilde_hat[i]
      }else{
        J[i,j] = -p_tilde_hat[i]
      }
    }
  }

  J = J/sum(p_tilde_hat)^2

  asyV = (t(J)%*%solve(Q)%*%Sigma%*%solve(Q)%*%t(J))/G

  p_hat = p_tilde_hat/sum(p_tilde_hat)
  p_est = rbind(p_est,c(p_hat))
  
  p_tilde_var = rbind(p_tilde_var,diag(solve(Q)%*%Sigma%*%solve(Q)))
  
  p_var = rbind(p_var,diag(asyV))

}

ci_l = p_est - qnorm(0.975)*sqrt(p_var)
ci_r = p_est + qnorm(0.975)*sqrt(p_var)

#coverage

coverage = ((rep(1,100)%*%t(bulk_p))&gt;=ci_l) &amp; ((rep(1,100)%*%t(bulk_p))&lt;=ci_r)
apply(coverage,2,mean)</code></pre>
<pre><code>[1] 0.93 0.96 0.95 0.98</code></pre>
<pre class="r"><code>#for(i in 1:4){
#  plot(ci_l[,i],type=&#39;l&#39;,ylim=range(c(ci_l[,i],ci_r[,i])))
#  lines(ci_r[,i],col=&#39;grey80&#39;)
#  lines(rep(bulk_p[i],100),lty = 2)
#}</code></pre>
<p>The bigger the beta, the lower the coverage??</p>
</div>
<div id="pbmc-data" class="section level2">
<h2>pbmc DATA</h2>
<pre class="r"><code>load(&quot;~/misc/data/scde/scCD14.RData&quot;)
load(&quot;~/misc/data/scde/scCD4.RData&quot;)
load(&quot;~/misc/data/scde/scCD8.RData&quot;)
#load(&quot;~/misc/data/scde/scCDT.RData&quot;)
load(&quot;~/misc/data/scde/scMB.RData&quot;)

# remove genes 
CD14 = as.matrix(CD14)
CD4 = as.matrix(CD4)
CD8 = as.matrix(CD8)
#CDT = as.matrix(CDT)
MB = as.matrix(MB)
rm.idx1 = which(((rowSums(CD14!=0)&lt;=10)|(rowSums(CD14)&gt;=quantile(rowSums(CD14),0.95))))
rm.idx2 = which(((rowSums(CD4!=0)&lt;=10)|(rowSums(CD4)&gt;=quantile(rowSums(CD4),0.95))))
rm.idx3 = which(((rowSums(CD8!=0)&lt;=10)|(rowSums(CD8)&gt;=quantile(rowSums(CD8),0.95))))
#rm.idx4 = which(((rowSums(CDT!=0)&lt;=10)|(rowSums(CDT)&gt;=quantile(rowSums(CDT),0.95))))
rm.idx5 = which(((rowSums(MB!=0)&lt;=10)|(rowSums(MB)&gt;=quantile(rowSums(MB),0.95))))

rm.idx = unique(c(rm.idx1,rm.idx2,rm.idx3,rm.idx5))

CD14 = CD14[-rm.idx,]
CD4 = CD4[-rm.idx,]
CD8 = CD8[-rm.idx,]
#CDT = CDT[-rm.idx,]
MB = MB[-rm.idx,]

Theta = cbind(rowSums(CD14)/sum(CD14),
              rowSums(CD4)/sum(CD4),
              rowSums(CD8)/sum(CD8),
              #rowSums(CDT)/sum(CDT),
              rowSums(MB)/sum(MB))

cor(Theta)</code></pre>
<pre><code>          [,1]      [,2]      [,3]      [,4]
[1,] 1.0000000 0.8534940 0.7874808 0.6553133
[2,] 0.8534940 1.0000000 0.7390032 0.6612358
[3,] 0.7874808 0.7390032 1.0000000 0.5953219
[4,] 0.6553133 0.6612358 0.5953219 1.0000000</code></pre>
<pre class="r"><code>kappa(t(Theta)%*%Theta)</code></pre>
<pre><code>[1] 57.73301</code></pre>
<pre class="r"><code># cell library size
S = c(sum(CD14)/ncol(CD14),
      sum(CD4)/ncol(CD4),
      sum(CD8)/ncol(CD8),
      sum(MB)/ncol(MB))</code></pre>
<pre class="r"><code>G = nrow(Theta)
K = ncol(Theta)

set.seed(12345)
# bulk data library size: 300*number of genes.
# bulk_ls = rpois(1,300*nrow(Theta))
# total number of cells in bulk data
bulk_ncell = 400
# cell proportions
bulk_beta = c(1,2,3,4)
bulk_beta = bulk_beta/sum(bulk_beta)
bulk_p = (bulk_beta * (S))/sum(bulk_beta * (S))
# bulk data gene relative expression.
bulk_X = bulk_ncell*Theta%*%diag(S)%*%bulk_beta
bulk_theta = bulk_X/sum(bulk_X)

ref_ncell = 100
ref_X = Theta%*%diag(S)*ref_ncell

w = rep(1,G)


p_tilde_est = c()
p_tilde_var = c()
p_est = c()
p_var = c()

nrep = 100
for(rep in 1:nrep){
  if(rep%%200==0){print(rep)}
  bulk_ls = rpois(1,300*nrow(Theta))
  Z = matrix(rpois(G*K,ref_X),ncol=K)
  y = rpois(G,bulk_ls*bulk_theta)

  #W = diag(w)
  #beta_hat = solve(t(Z)%*%W%*%Z - diag(colSums(W%*%Z)))%*%t(Z)%*%W%*%y

  U = diag(colSums(Z))
  Zu = Z%*%(solve(U))
  Zuu = Zu%*%solve(U)
  p_tilde_hat = solve(t(Zu)%*%Zu - diag(colSums(Zuu)))%*%t(Zu)%*%y
  p_tilde_est = rbind(p_tilde_est,c(p_tilde_hat))
  #beta_hat = solve(t(Z)%*%Z)%*%t(Z)%*%y

  #beta_hat
  #beta_hat/sum(beta_hat)

  Q = 0
  Sigma=0
  deltas = c()
  for(i in 1:length(y)){ 
    ag = Zu[i,]%*%t(Zu[i,])-diag(Zuu[i,])
    Q = Q + ag
    Delta = (ag%*%p_tilde_hat-y[i]*Zu[i,])
    Sigma = Sigma + Delta%*%t(Delta)
  }
  Q = Q/G
  Sigma = Sigma/G

  J = matrix(nrow = K,ncol=K)
  for(i in 1:K){
    for(j in 1:K){
      if(i==j){
        J[i,j] = sum(p_tilde_hat)-p_tilde_hat[i]
      }else{
        J[i,j] = -p_tilde_hat[i]
      }
    }
  }

  J = J/sum(p_tilde_hat)^2

  asyV = ((J)%*%solve(Q)%*%Sigma%*%solve(Q)%*%t(J))/G

  p_hat = p_tilde_hat/sum(p_tilde_hat)
  p_est = rbind(p_est,c(p_hat))
  
  p_tilde_var = rbind(p_tilde_var,diag(solve(Q)%*%Sigma%*%solve(Q)))
  
  p_var = rbind(p_var,diag(asyV))

}

ci_l = p_est - qnorm(0.975)*sqrt(p_var)
ci_r = p_est + qnorm(0.975)*sqrt(p_var)

#coverage

coverage = ((rep(1,100)%*%t(bulk_p))&gt;=ci_l) &amp; ((rep(1,100)%*%t(bulk_p))&lt;=ci_r)
apply(coverage,2,mean)</code></pre>
<pre><code>[1] 0.96 0.98 0.94 0.98</code></pre>
<p>how about change <span class="math inline">\(beta\)</span> to be <span class="math inline">\((0.4,0.3,0.2,0.1)\)</span>?</p>
<pre class="r"><code>G = nrow(Theta)
K = ncol(Theta)

set.seed(12345)
# bulk data library size: 300*number of genes.
# bulk_ls = rpois(1,300*nrow(Theta))
# total number of cells in bulk data
bulk_ncell = 400
# cell proportions
bulk_beta = c(4,3,2,1)
bulk_beta = bulk_beta/sum(bulk_beta)
bulk_p = (bulk_beta * (S))/sum(bulk_beta * (S))
# bulk data gene relative expression.
bulk_X = bulk_ncell*Theta%*%diag(S)%*%bulk_beta
bulk_theta = bulk_X/sum(bulk_X)

ref_ncell = 100
ref_X = Theta%*%diag(S)*ref_ncell

w = rep(1,G)


p_tilde_est = c()
p_tilde_var = c()
p_est = c()
p_var = c()

nrep = 100
for(rep in 1:nrep){
  if(rep%%200==0){print(rep)}
  bulk_ls = rpois(1,300*nrow(Theta))
  Z = matrix(rpois(G*K,ref_X),ncol=K)
  y = rpois(G,bulk_ls*bulk_theta)

  #W = diag(w)
  #beta_hat = solve(t(Z)%*%W%*%Z - diag(colSums(W%*%Z)))%*%t(Z)%*%W%*%y

  U = diag(colSums(Z))
  Zu = Z%*%(solve(U))
  Zuu = Zu%*%solve(U)
  p_tilde_hat = solve(t(Zu)%*%Zu - diag(colSums(Zuu)))%*%t(Zu)%*%y
  p_tilde_est = rbind(p_tilde_est,c(p_tilde_hat))
  #beta_hat = solve(t(Z)%*%Z)%*%t(Z)%*%y

  #beta_hat
  #beta_hat/sum(beta_hat)

  Q = 0
  Sigma=0
  deltas = c()
  for(i in 1:length(y)){ 
    ag = Zu[i,]%*%t(Zu[i,])-diag(Zuu[i,])
    Q = Q + ag
    Delta = (ag%*%p_tilde_hat-y[i]*Zu[i,])
    Sigma = Sigma + Delta%*%t(Delta)
  }
  Q = Q/G
  Sigma = Sigma/G

  J = matrix(nrow = K,ncol=K)
  for(i in 1:K){
    for(j in 1:K){
      if(i==j){
        J[i,j] = sum(p_tilde_hat)-p_tilde_hat[i]
      }else{
        J[i,j] = -p_tilde_hat[i]
      }
    }
  }

  J = J/sum(p_tilde_hat)^2

  asyV = ((J)%*%solve(Q)%*%Sigma%*%solve(Q)%*%t(J))/G

  p_hat = p_tilde_hat/sum(p_tilde_hat)
  p_est = rbind(p_est,c(p_hat))
  
  p_tilde_var = rbind(p_tilde_var,diag(solve(Q)%*%Sigma%*%solve(Q)))
  
  p_var = rbind(p_var,diag(asyV))

}

ci_l = p_est - qnorm(0.975)*sqrt(p_var)
ci_r = p_est + qnorm(0.975)*sqrt(p_var)

#coverage

coverage = ((rep(1,100)%*%t(bulk_p))&gt;=ci_l) &amp; ((rep(1,100)%*%t(bulk_p))&lt;=ci_r)
apply(coverage,2,mean)</code></pre>
<pre><code>[1] 0.98 0.95 0.95 0.96</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.5.1 (2018-07-02)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Scientific Linux 7.4 (Nitrogen)

Matrix products: default
BLAS/LAPACK: /software/openblas-0.2.19-el7-x86_64/lib/libopenblas_haswellp-r0.2.19.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
[1] Matrix_1.2-15       Biobase_2.42.0      BiocGenerics_0.28.0

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.2      knitr_1.20      whisker_0.3-2   magrittr_1.5   
 [5] workflowr_1.6.0 lattice_0.20-38 R6_2.3.0        stringr_1.3.1  
 [9] highr_0.7       tools_3.5.1     grid_3.5.1      git2r_0.26.1   
[13] htmltools_0.3.6 yaml_2.2.0      rprojroot_1.3-2 digest_0.6.18  
[17] later_0.7.5     promises_1.0.1  fs_1.3.1        glue_1.3.0     
[21] evaluate_0.12   rmarkdown_1.10  stringi_1.2.4   compiler_3.5.1 
[25] backports_1.1.2 httpuv_1.4.5   </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
