<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="DongyueXie" />

<meta name="date" content="2020-02-17" />

<title>Single cell cytokines data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.0.13/css/fa-svg-with-js.css" rel="stylesheet" />
<script src="site_libs/font-awesome-5.0.13/js/fontawesome-all.min.js"></script>
<script src="site_libs/font-awesome-5.0.13/js/fa-v4-shims.min.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">misc</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/DongyueXie/misc">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Single cell cytokines data</h1>
<h4 class="author"><em>DongyueXie</em></h4>
<h4 class="date"><em>2020-02-17</em></h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-02-26
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>misc/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20191122code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20191122)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20191122code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20191122)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongabsolute"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>File paths:</strong> absolute </a>
</p>
</div>
<div id="strongFilepathsstrongabsolute" class="panel-collapse collapse">
<div class="panel-body">
<p>
Using absolute paths to the files within your workflowr project makes it difficult for you and others to run your code on a different machine. Change the absolute path(s) below to the suggested relative path(s) to make your code more reproducible.
</p>
<table class="table table-condensed table-hover">
<thead>
<tr>
<th style="text-align:left;">
absolute
</th>
<th style="text-align:left;">
relative
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
~/misc/data/cytokine_normalized.RData
</td>
<td style="text-align:left;">
data/cytokine_normalized.RData
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomDongyueXiemisctreeca23be2ba1c86473f2e784c0f94a6ff4f9f1d334targetblankca23be2a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/DongyueXie/misc/tree/ca23be2ba1c86473f2e784c0f94a6ff4f9f1d334" target="_blank">ca23be2</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomDongyueXiemisctreeca23be2ba1c86473f2e784c0f94a6ff4f9f1d334targetblankca23be2a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  analysis/methylation.Rmd
    Untracked:  code/Rplots.pdf
    Untracked:  code/sccytokines.R
    Untracked:  data/Banknote_result.RData
    Untracked:  data/BreastCancer_result.RData
    Untracked:  data/Haberman_result.RData
    Untracked:  data/Ionosphere_result.RData
    Untracked:  data/Mice_result.RData
    Untracked:  data/PimaIndiansDiabetes_result.RData
    Untracked:  data/ROTS_results.RData
    Untracked:  data/Seeds_result.RData
    Untracked:  data/Sonar_result.RData
    Untracked:  data/cytokine/
    Untracked:  data/cytokine_normalized.RData
    Untracked:  data/liver_result.RData

Unstaged changes:
    Modified:   analysis/index.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DongyueXie/misc/blob/ca23be2ba1c86473f2e784c0f94a6ff4f9f1d334/analysis/sccytokines.Rmd" target="_blank">ca23be2</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-02-26
</td>
<td>
wflow_publish(“analysis/sccytokines.Rmd”)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>See if sva works when comparing two control groups in single cell cytokine study.</p>
<p>Take a look at the annotation file:</p>
<pre class="r"><code>data_ann = read.csv(&#39;/project2/mstephens/chevrier-stephens/data/whole_cyto_annot.csv.gz&#39;)
data_ann[1:5,]</code></pre>
<pre><code>                  X0 sample well n_counts log_counts n_genes     mt_frac
1 AAACCTGAGAATAGGG-1  CCL20    1     1611   7.384610     852 0.030415890
2 AAACCTGAGAGTAAGG-1  CXCL1    1     6482   8.776784    2340 0.032860227
3 AAACCTGAGTGGCACA-1  CCL20    1     1940   7.570443    1058 0.048453607
4 AAACCTGCAAGCTGGA-1  CXCL1    1     2186   7.689829    1014 0.030192131
5 AAACCTGCAAGGACAC-1  CCL20    1      849   6.744059     469 0.001177856
  louvain louvain_sub_5 louvain_sub_9 subclustered   cell_type
1       0             0             0            0     B_cells
2       0             0             0            0     B_cells
3       1             1             1            1     B_cells
4       0             0             0            0     B_cells
5       3             3             3            3 Neutrophils</code></pre>
<p>sample: treatment or cytokines, including two controls.</p>
<p>well? There are 25 wells, each of which contains two samples.</p>
<p>cell_type: focus on 8 types of cells, B_cells, CD4_T_cells, CD8_T_cells, NK_cells, Dendritic-cells, Ly6C+, Ly6C-,Neutrolphils.</p>
<pre class="r"><code>#read data matrix
# library(hdf5r)
# library(Matrix)
# f &lt;- H5File$new(&quot;whole_cyto_normalized.h5ad&quot;,mode = &quot;r&quot;)
# print(names(f))
# out &lt;- f[[&quot;X&quot;]]
# print(h5attributes(out))
# i &lt;- out[[&quot;indices&quot;]][]
# j &lt;- out[[&quot;indptr&quot;]][]
# x &lt;- out[[&quot;data&quot;]][]
#library(SparseM)
# X.csr = new(&#39;matrix.csr&#39;,ra=x,ja=as.integer(i+1),ia=as.integer(j+1),dimension=h5attributes(out)$h5sparse_shape)
#load(&quot;~/misc/data/cytokine_normalized.RData&quot;)
load(&#39;data/cytokine/DE_controls_output.RData&#39;)</code></pre>
<p>Number of cells in each control group:</p>
<pre class="r"><code>n.sample = c()
for(cell in names(output)){
  n.sample = rbind(n.sample,c(sum(output[[cell]]$group_idx),sum(1-output[[cell]]$group_idx)))
}
rownames(n.sample) = names(output)
colnames(n.sample) = c(&#39;Ctrl-1&#39;,&#39;Ctrl-2&#39;)
knitr::kable(n.sample,caption = &#39;Number of samples&#39;)</code></pre>
<table>
<caption>Number of samples</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">Ctrl-1</th>
<th align="right">Ctrl-2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>B_cells</td>
<td align="right">2326</td>
<td align="right">2348</td>
</tr>
<tr class="even">
<td>CD4_T_cells</td>
<td align="right">188</td>
<td align="right">318</td>
</tr>
<tr class="odd">
<td>CD8_T_cells</td>
<td align="right">291</td>
<td align="right">448</td>
</tr>
<tr class="even">
<td>NK_cells</td>
<td align="right">36</td>
<td align="right">32</td>
</tr>
<tr class="odd">
<td>Dendritic_cells</td>
<td align="right">26</td>
<td align="right">15</td>
</tr>
<tr class="even">
<td>Ly6C+_Monocytes</td>
<td align="right">75</td>
<td align="right">89</td>
</tr>
<tr class="odd">
<td>Ly6C-_Monocytes</td>
<td align="right">128</td>
<td align="right">92</td>
</tr>
<tr class="even">
<td>Neutrophils</td>
<td align="right">227</td>
<td align="right">188</td>
</tr>
</tbody>
</table>
<p>Number of genes that have at least one measurement in two control groups.</p>
<pre class="r"><code>n.gene = c(14853)
for(cell in names(output)){
  n.gene = rbind(n.gene,c(14853-length(output[[cell]]$rm.idx)))
}
rownames(n.gene) = c(&#39;Total&#39;,names(output))
knitr::kable(n.gene,caption = &#39;Number of genes considered&#39;,col.names = &#39;#genes&#39;)</code></pre>
<table>
<caption>Number of genes considered</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">#genes</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Total</td>
<td align="right">14853</td>
</tr>
<tr class="even">
<td>B_cells</td>
<td align="right">13130</td>
</tr>
<tr class="odd">
<td>CD4_T_cells</td>
<td align="right">10974</td>
</tr>
<tr class="even">
<td>CD8_T_cells</td>
<td align="right">11565</td>
</tr>
<tr class="odd">
<td>NK_cells</td>
<td align="right">8816</td>
</tr>
<tr class="even">
<td>Dendritic_cells</td>
<td align="right">9512</td>
</tr>
<tr class="odd">
<td>Ly6C+_Monocytes</td>
<td align="right">10568</td>
</tr>
<tr class="even">
<td>Ly6C-_Monocytes</td>
<td align="right">10962</td>
</tr>
<tr class="odd">
<td>Neutrophils</td>
<td align="right">8659</td>
</tr>
</tbody>
</table>
</div>
<div id="correlations-between-pcs-and-groups" class="section level2">
<h2>Correlations between PCs and groups</h2>
<p>Plot of correlations between groups and first 20 principle components.</p>
<pre class="r"><code>par(mfrow=c(3,3))
for(cell in names(output)){
  plot(output[[cell]]$pc.cor[1:20],xlab=&#39;PCs&#39;,ylab=&#39;corr&#39;,main=paste(cell),ylim = c(-0.5,0.6),pch=20)
  abline(h=0,lty=3)
}</code></pre>
<p><img src="figure/sccytokines.Rmd/unnamed-chunk-5-1.png" width="864" style="display: block; margin: auto;" /></p>
<p>Plot of principle component that has maximum absolute correlation with groups for each cell. Vertical line seperates two groups.</p>
<pre class="r"><code>par(mfrow=c(3,3))
for(cell in names(output)){
  plot(output[[cell]]$pc.cor.max,xlab=&#39;&#39;,ylab=&#39;&#39;,
       main=paste(cell,&#39;, PC:&#39;,which.max(abs(output[[cell]]$pc.cor)),&#39;, corr:&#39;,round(max(abs(output[[cell]]$pc.cor)),2),sep=&#39;&#39;),pch=1,col=&#39;grey50&#39;)
  abline(v=sum(output[[cell]]$group_idx),lty=3)
}</code></pre>
<p><img src="figure/sccytokines.Rmd/unnamed-chunk-6-1.png" width="864" style="display: block; margin: auto;" /></p>
</div>
<div id="compare-p-value-distributions-t-test-and-sva-limma" class="section level2">
<h2>Compare p-value distributions, t-test and sva-limma</h2>
<p>Note: We will see a lot of p-values from t-test around 0.3-0.4. In single cell DE study, some genes are only measured once in one group while have no observation in another group. For example, gene expression in group 1 <span class="math inline">\(= (0,0,...,0,0)\)</span> and gene expression in group 2 <span class="math inline">\(= (x,0,...,0,0)\)</span>. So in this case, unequal variance two-sample t-test always gives t-statistic = <span class="math inline">\(1\)</span> with <span class="math inline">\(df=n_2-1\)</span>, where <span class="math inline">\(n_2\)</span> is the number of samples in group 2. Let’s plot p-value vs df.</p>
<p>Suppose we have at least 5 samples in group 2, then p-value starts at 0.3739(df=4) and converges to 0.3173 as df goes to infinite.</p>
<pre class="r"><code>par(mfrow=c(1,1))
plot(4:1e3,(1-pt(1,4:1e3))*2,xlab=&#39;df&#39;,ylab=&#39;p-value&#39;,main=&#39;t-statistics = 1&#39;,pch=20)</code></pre>
<p><img src="figure/sccytokines.Rmd/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Now compare distributions of p-values from t-test and sva-limma:</p>
<pre class="r"><code>par(mfrow=c(8,2))
for(cell in names(output)){
  hist(output[[cell]]$pvalue_t,main=paste(cell, &#39;,t test&#39;),xlab=&#39;&#39;)
  hist(output[[cell]]$pvalue_sva_limma,main=paste(cell, &#39;,sva-limma&#39;),xlab=&#39;&#39;)
}</code></pre>
<p><img src="figure/sccytokines.Rmd/unnamed-chunk-8-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<div id="compare-the-number-of-significant-genes" class="section level2">
<h2>Compare the number of significant genes</h2>
<p>The number of significant genes at <span class="math inline">\(fdr=0.05\)</span> by BH procedure.</p>
<pre class="r"><code>par(mfrow=c(1,1))
n.sig = c()
for(cell in names(output)){
  n.sig = rbind(n.sig,c(length(output[[cell]]$rej.idx.ttest),length(output[[cell]]$rej.idx.sva)))
}
rownames(n.sig) = names(output)
colnames(n.sig) = c(&#39;t-test&#39;,&#39;sva&#39;)
knitr::kable(n.sig,caption = &#39;Number of significant genes at fdr=0.05&#39;)</code></pre>
<table>
<caption>Number of significant genes at fdr=0.05</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">t-test</th>
<th align="right">sva</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>B_cells</td>
<td align="right">859</td>
<td align="right">839</td>
</tr>
<tr class="even">
<td>CD4_T_cells</td>
<td align="right">25</td>
<td align="right">11</td>
</tr>
<tr class="odd">
<td>CD8_T_cells</td>
<td align="right">101</td>
<td align="right">61</td>
</tr>
<tr class="even">
<td>NK_cells</td>
<td align="right">2</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td>Dendritic_cells</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td>Ly6C+_Monocytes</td>
<td align="right">22</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td>Ly6C-_Monocytes</td>
<td align="right">16</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td>Neutrophils</td>
<td align="right">1</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<p>The number of significant genes at <span class="math inline">\(fdr=0.01\)</span> by BH procedure.</p>
<pre class="r"><code>BH = function(p,alpha=0.05){
  n=length(p)
  i=rank(p)
  idx = which(p&lt;=(i/n*alpha))
  if(length(idx)==0){
    NULL
  }else{
    i0= max(i[idx])
    rej.idx = which(i&lt;=i0)
    rej.idx
  }
}

n.sig = c()
for(cell in names(output)){
  n.sig = rbind(n.sig,c(length(BH(output[[cell]]$pvalue_t,0.01)),length(BH(output[[cell]]$pvalue_sva_limma,0.01))))
}
rownames(n.sig) = names(output)
colnames(n.sig) = c(&#39;t-test&#39;,&#39;sva&#39;)
knitr::kable(n.sig,caption = &#39;Number of significant genes at fdr=0.01&#39;)</code></pre>
<table>
<caption>Number of significant genes at fdr=0.01</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">t-test</th>
<th align="right">sva</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>B_cells</td>
<td align="right">463</td>
<td align="right">465</td>
</tr>
<tr class="even">
<td>CD4_T_cells</td>
<td align="right">18</td>
<td align="right">7</td>
</tr>
<tr class="odd">
<td>CD8_T_cells</td>
<td align="right">51</td>
<td align="right">44</td>
</tr>
<tr class="even">
<td>NK_cells</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td>Dendritic_cells</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td>Ly6C+_Monocytes</td>
<td align="right">13</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td>Ly6C-_Monocytes</td>
<td align="right">11</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td>Neutrophils</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.5.1 (2018-07-02)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Scientific Linux 7.4 (Nitrogen)

Matrix products: default
BLAS/LAPACK: /software/openblas-0.2.19-el7-x86_64/lib/libopenblas_haswellp-r0.2.19.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] workflowr_1.6.0 Rcpp_1.0.2      digest_0.6.18   later_0.7.5    
 [5] rprojroot_1.3-2 R6_2.3.0        backports_1.1.2 git2r_0.26.1   
 [9] magrittr_1.5    evaluate_0.12   highr_0.7       stringi_1.2.4  
[13] fs_1.3.1        promises_1.0.1  whisker_0.3-2   rmarkdown_1.10 
[17] tools_3.5.1     stringr_1.3.1   glue_1.3.0      httpuv_1.4.5   
[21] yaml_2.2.0      compiler_3.5.1  htmltools_0.3.6 knitr_1.20     </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
