<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="DongyueXie" />

<meta name="date" content="2020-01-15" />

<title>scde using PBMC data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">misc</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/DongyueXie/misc">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">scde using PBMC data</h1>
<h4 class="author">DongyueXie</h4>
<h4 class="date">2020-01-15</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-02-28
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>misc/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.5.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20191122code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20191122)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20191122code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20191122)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomDongyueXiemisctree327dcd2e516ffc90a0256595524cdfe4edadf758targetblank327dcd2a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/DongyueXie/misc/tree/327dcd2e516ffc90a0256595524cdfe4edadf758" target="_blank">327dcd2</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomDongyueXiemisctree327dcd2e516ffc90a0256595524cdfe4edadf758targetblank327dcd2a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  data/ROTS_results.RData
    Untracked:  data/ROTS_results2.RData
    Untracked:  data/cate_cate_high.RData
    Untracked:  data/cate_cate_low.RData
    Untracked:  data/cate_cate_mid.RData
    Untracked:  data/cate_cate_null.RData
    Untracked:  data/mout_high.RData
    Untracked:  data/mout_low.RData
    Untracked:  data/mout_mid.RData
    Untracked:  data/mout_null.RData
    Untracked:  data/pbmc.rds
    Untracked:  data/pbmc_counts.rds
    Untracked:  data/sva_sva_high.RData
    Untracked:  data/sva_sva_low.RData
    Untracked:  data/sva_sva_mid.RData
    Untracked:  data/sva_sva_null.RData

Unstaged changes:
    Modified:   analysis/deconvolution.Rmd
    Modified:   analysis/index.Rmd
    Modified:   analysis/scdePBMC.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DongyueXie/misc/blob/327dcd2e516ffc90a0256595524cdfe4edadf758/analysis/scdePBMC2.Rmd" target="_blank">327dcd2</a>
</td>
<td>
Dongyue Xie
</td>
<td>
2020-02-28
</td>
<td>
wflow_publish(“analysis/scdePBMC2.Rmd”)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>In my previous <a href="scde.html">analysis</a>, I tried RUV methods on single-cell RNA-Seq data. I only tried top 1000 expressed genes from the dataset. So only a few of gene expressions are 0. This is not what typically scRNA-Seq data are. Also, I tried only 50 cells.</p>
<p>This time, I’m going to try PBMC data prepared by <a href="https://satijalab.org/seurat/v3.1/pbmc3k_tutorial.html">Satjia Lab</a>.</p>
<pre class="r"><code>library(MAST)
library(Seurat)
datax = readRDS(&#39;data/pbmc_counts.rds&#39;)@assays$RNA
datax = datax[rowSums(datax)&gt;0,]
clusters = readRDS(&#39;data/pbmc.rds&#39;)

cell_cluster = clusters@colData$seurat</code></pre>
<p>How many zeros are there? A lot</p>
<pre class="r"><code>sum(datax==0)/(dim(datax)[1]*dim(datax)[2])</code></pre>
<pre><code>[1] 0.9381137</code></pre>
<p>Let’s only use the Naive CD4+ T cells, which corresponds to the first cluster. Total 709 cells.</p>
<pre class="r"><code>CDT_idx = which(cell_cluster == 1)
CDT = datax[,CDT_idx]

set.seed(12345)
group1_idx = sample(1:ncol(CDT),ncol(CDT)/2)
group1 = CDT[,group1_idx]
group2 = CDT[,-group1_idx]
## for each gene, run a two-sample t test

p_values1 = c()
for(i in 1:nrow(CDT)){
  p_values1[i] = t.test(log(group1[i,]+1),log(group2[i,]+1),alternative=&#39;two.sided&#39;)$p.value
}
hist(p_values1,breaks = 15)</code></pre>
<p><img src="figure/scdePBMC2.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>summary(p_values1)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA&#39;s 
 0.0001  0.2539  0.4368  0.4842  0.7154  0.9997    1225 </code></pre>
<p>There are 1225 genes that have no expression in any Naive CD4+ T cells. Remove these genes.</p>
<pre class="r"><code>CDT = CDT[-which((rowSums(CDT)==0)),]</code></pre>
<p>Again, t-test does not apply here since for a lot of genes, only one cell has non-zero read counts among two groups.</p>
<p>Now let’s apply ROTS method for scDE, which is one of the best performance method tested in <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-91">Sonenson and Delorenzi</a>.</p>
<pre class="r"><code># First Normalize the counts by Trimmed Mean of M-values as required by ROTS. Then apply ROTS.
group = rep(0,ncol(CDT))
group[group1_idx] = 1
library(ROTS)
#ROTS_results = ROTS(data = CDTnorm, groups = group , B = 100 , K = 500 , seed = 1234)
load(&#39;data/ROTS_results.RData&#39;)
summary(ROTS_results, fdr = 0.05)</code></pre>
<pre><code>ROTS results: 

Number of resamplings:  100 

a1:                     4.6 
a2:                     1 
Top list size:          390 
Reproducibility value:  0.1662051 
Z-score:                5.651558 

0 rows satisfy the condition.
     Row ROTS-statistic pvalue FDR</code></pre>
<pre class="r"><code>hist(ROTS_results$pvalue,breaks = 15)</code></pre>
<p><img src="figure/scdePBMC2.Rmd/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="ruv-methods" class="section level2">
<h2>RUV methods</h2>
<p>First apply on NULL data then add signals to genes using Poisson thinning.</p>
<p>Randomly split 709 cells to two groups.</p>
<pre class="r"><code>library(vicar)

set.seed(12345)
group1_idx = sample(1:ncol(CDT),ncol(CDT)/2)
group1 = CDT[,group1_idx]
group2 = CDT[,-group1_idx]

group_indicator = rep(0,ncol(CDT))
group_indicator[group1_idx] = 1

X = model.matrix(~group_indicator)
CDT = as.matrix(CDT)

Y = t((CDT))

#num_sv     &lt;- sva::num.sv(dat = t(Y), mod = X, method = &quot;be&quot;)
#num_sv_l   &lt;- sva::num.sv(dat = t(Y), mod = X, method = &quot;leek&quot;)

num_sv = 3
#num_sv_l</code></pre>
<p>The number of estimated surrogate variables is 3.</p>
<pre class="r"><code>eps=0.01
Y = log(Y+eps)
mout = mouthwash(Y,X,k=num_sv,cov_of_interest = 2,include_intercept = FALSE)</code></pre>
<pre><code>Running mouthwash on 709 x 2 matrix X and 709 x 12488 matrix Y.
 - Computing independent basis using QR decomposition.
 - Computation took 23.891 seconds.
 - Running additional preprocessing steps.
 - Computation took 0.002 seconds.
 - Running second step of mouthwash:
    + Estimating model parameters using EM.
    + Computation took 134.018 seconds.
    + Generating adaptive shrinkage (ash) output.
    + Computation took 1.46 seconds.
 - Second step took 136.331 seconds.
 - Estimating additional hidden confounders.
 - Computation took 2.991 seconds.</code></pre>
<pre class="r"><code>save(mout,file = &#39;data/mout_null.RData&#39;)
#load(&#39;data/mout_null.RData&#39;)
mout$pi0</code></pre>
<pre><code>[1] 0.9991927</code></pre>
<pre class="r"><code>library(cate)
#library(leapp)

cate_cate   &lt;- cate::cate.fit(X.primary = X[, 2, drop = FALSE], X.nuis = X[, -2, drop = FALSE],
                              Y = Y, r = num_sv, adj.method = &quot;rr&quot;)
save(cate_cate,file = &#39;data/cate_cate_null.RData&#39;)
#load(&#39;data/cate_cate_null.RData&#39;)
# this method is vey slow!
#leapp_leapp &lt;- leapp::leapp(data = t(Y), pred.prim = X[, 2, drop = FALSE], 
#                            pred.covar = X[, -2, drop = FALSE], num.fac = num_sv)

sva_sva     &lt;- sva::sva(dat = t(Y), mod = X, mod0 = X[, -2, drop = FALSE], n.sv = num_sv)</code></pre>
<pre><code>Number of significant surrogate variables is:  3 
Iteration (out of 5 ):1  2  3  4  5  </code></pre>
<pre class="r"><code>save(sva_sva,file = &#39;data/sva_sva_null.RData&#39;)
#load(&#39;data/sva_sva_null.RData&#39;)
X.sva &lt;- cbind(X, sva_sva$sv)
lmout &lt;- limma::lmFit(object = t(Y), design = X.sva)
eout  &lt;- limma::eBayes(lmout)
svaout           &lt;- list()
svaout$betahat   &lt;- lmout$coefficients[, 2]
svaout$sebetahat &lt;- lmout$stdev.unscaled[, 2] * sqrt(eout$s2.post)
svaout$pvalues   &lt;- eout$p.value[, 2]

hist(svaout$pvalues,breaks=15)</code></pre>
<p><img src="figure/scdePBMC2.Rmd/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ks.test(svaout$pvalues,&#39;punif&#39;,0,1)</code></pre>
<pre><code>
    One-sample Kolmogorov-Smirnov test

data:  svaout$pvalues
D = 0.074756, p-value &lt; 2.2e-16
alternative hypothesis: two-sided</code></pre>
<pre class="r"><code>hist(cate_cate$beta.p.value,breaks = 15)</code></pre>
<p><img src="figure/scdePBMC2.Rmd/unnamed-chunk-10-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ks.test(cate_cate$beta.p.value,&#39;punif&#39;,0,1)</code></pre>
<pre><code>
    One-sample Kolmogorov-Smirnov test

data:  cate_cate$beta.p.value
D = 0.072026, p-value &lt; 2.2e-16
alternative hypothesis: two-sided</code></pre>
<p>Add some signal to the NULL dataset.</p>
<pre class="r"><code>library(seqgendiff)
#tt = thin_diff(round(cell16), design_fixed = X[,2,drop=FALSE])
set.seed(12345)
thinout = thin_2group(round(CDT),0.9,signal_fun = stats::rexp,signal_params = list(rate=0.5))

#check null groups

group1 = CDT[,which(thinout$designmat==1)]
group2 = CDT[,which(thinout$designmat==0)]
## for each gene, run a two-sample t test

#p_values1 = c()
#for(i in 1:nrow(CDT)){
#  p_values1[i] = t.test(group1[i,],group2[i,],alternative=&#39;two.sided&#39;)$p.value
#}
#ks.test(p_values1,&#39;punif&#39;,0,1)
#hist(p_values1,breaks = 15)

Y = t(thinout$mat)

remove.idx = which(colSums(Y)==0)

Y = log(Y[,-remove.idx]+eps)

X = model.matrix(~thinout$designmat)

#num_sv     &lt;- sva::num.sv(dat = t(Y), mod = X, method = &quot;be&quot;)
#num_sv_l   &lt;- sva::num.sv(dat = t(Y), mod = X, method = &quot;leek&quot;)

num_sv = 3
#num_sv_l

mean(abs(thinout$coef) &lt; 10^-6)</code></pre>
<pre><code>[1] 0.899984</code></pre>
<pre class="r"><code>mout = mouthwash(Y,X,k=num_sv,cov_of_interest = 2,include_intercept = FALSE)</code></pre>
<pre><code>Running mouthwash on 709 x 2 matrix X and 709 x 12446 matrix Y.
 - Computing independent basis using QR decomposition.
 - Computation took 22.124 seconds.
 - Running additional preprocessing steps.
 - Computation took 0 seconds.
 - Running second step of mouthwash:
    + Estimating model parameters using EM.
    + Computation took 179.905 seconds.
    + Generating adaptive shrinkage (ash) output.
    + Computation took 1.419 seconds.
 - Second step took 182.303 seconds.
 - Estimating additional hidden confounders.
 - Computation took 2.433 seconds.</code></pre>
<pre class="r"><code>save(mout,file = &#39;data/mout_mid.RData&#39;)
mout$pi0</code></pre>
<pre><code>[1] 0.9822051</code></pre>
<pre class="r"><code>#load(&#39;data/mout_mid.RData&#39;)

#bout &lt;- backwash(Y = Y, X = X, k = num_sv, cov_of_interest = 2, include_intercept = FALSE)
#save(bout,file = &#39;data/bout_mid.RData&#39;)
#bout$pi0


cate_cate = cate::cate.fit(X.primary = X[, 2, drop = FALSE], X.nuis = X[, -2, drop = FALSE],
                              Y = Y, r = num_sv, adj.method = &quot;rr&quot;)
save(cate_cate,file = &#39;data/cate_cate_mid.RData&#39;)
#load(&#39;data/cate_cate_mid.RData&#39;)
sva_sva     &lt;- sva::sva(dat = t(Y), mod = X, mod0 = X[, -2, drop = FALSE], n.sv = num_sv)</code></pre>
<pre><code>Number of significant surrogate variables is:  3 
Iteration (out of 5 ):1  2  3  4  5  </code></pre>
<pre class="r"><code>save(sva_sva,file = &#39;data/sva_sva_mid.RData&#39;)
#load(&#39;data/sva_sva_mid.RData&#39;)
X.sva &lt;- cbind(X, sva_sva$sv)
lmout &lt;- limma::lmFit(object = t(Y), design = X.sva)
eout  &lt;- limma::eBayes(lmout)
svaout           &lt;- list()
svaout$betahat   &lt;- lmout$coefficients[, 2]
svaout$sebetahat &lt;- lmout$stdev.unscaled[, 2] * sqrt(eout$s2.post)
svaout$pvalues   &lt;- eout$p.value[, 2]

which_null = c(1*(abs(thinout$coef[-remove.idx]) &lt; 10^-6))



# plot ROC curve
roc_out &lt;- list(
  pROC::roc(response = which_null, predictor = c(mout$result$lfdr)),
  #pROC::roc(response = which_null, predictor = c(bout$result$lfdr)),
  pROC::roc(response = which_null, predictor = c(cate_cate$beta.p.value)),
  pROC::roc(response = which_null, predictor = c(svaout$pvalues)))
name_vec &lt;- c(&quot;MOUTHWASH&quot;, &quot;CATErr&quot;, &quot;SVA&quot;)
names(roc_out) &lt;- name_vec

sout &lt;- lapply(roc_out, function(x) { data.frame(TPR = x$sensitivities, FPR = 1 - x$specificities)})
for (index in 1:length(sout)) {
  sout[[index]]$Method &lt;- name_vec[index]
}
longdat &lt;- do.call(rbind, sout)

shortdat &lt;- dplyr::filter(longdat, Method == &quot;MOUTHWASH&quot; |
                            Method == &quot;CATErr&quot; | Method == &quot;SVA&quot; | Method == &quot;LEAPP&quot;)
library(ggplot2)
ggplot(data = shortdat, mapping = aes(x = FPR, y = TPR, col = Method)) +
  geom_path() + theme_bw() + ggtitle(&quot;ROC Curves&quot;)</code></pre>
<p><img src="figure/scdePBMC2.Rmd/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>auc_vec &lt;- sapply(roc_out, FUN = function(x) { x$auc })
knitr::kable(sort(auc_vec, decreasing = TRUE), col.names = &quot;AUC&quot;, digits = 3)</code></pre>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">AUC</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>SVA</td>
<td align="right">0.743</td>
</tr>
<tr class="even">
<td>CATErr</td>
<td align="right">0.734</td>
</tr>
<tr class="odd">
<td>MOUTHWASH</td>
<td align="right">0.630</td>
</tr>
</tbody>
</table>
<pre class="r"><code># estimate pi0
method_list &lt;- list()
method_list$CATErr           &lt;- list()
method_list$CATErr$betahat   &lt;- c(cate_cate$beta)
method_list$CATErr$sebetahat &lt;- c(sqrt(cate_cate$beta.cov.row * c(cate_cate$beta.cov.col)) / sqrt(nrow(X)))

method_list$SVA             &lt;- list()
method_list$SVA$betahat     &lt;- c(svaout$betahat)
method_list$SVA$sebetahat   &lt;- c(svaout$sebetahat)

ashfit &lt;- lapply(method_list, FUN = function(x) { ashr::ash(x$betahat, x$sebetahat)})
api0 &lt;- sapply(ashfit, FUN = ashr::get_pi0)
api0 &lt;- c(api0, MOUTHWASH = mout$pi0)
#api0 &lt;- c(api0, BACKWASH = bout$pi0)

knitr::kable(sort(api0, decreasing = TRUE), col.names = &quot;Estimate of Pi0&quot;)</code></pre>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">Estimate of Pi0</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>MOUTHWASH</td>
<td align="right">0.9822051</td>
</tr>
<tr class="even">
<td>SVA</td>
<td align="right">0.9766238</td>
</tr>
<tr class="odd">
<td>CATErr</td>
<td align="right">0.1653960</td>
</tr>
</tbody>
</table>
<p>Stronger signal: rexp(,rate = 0.2)</p>
<pre class="r"><code>set.seed(12345)
thinout = thin_2group(round(CDT),0.9,signal_fun = stats::rexp,signal_params = list(rate=0.2))



Y = t(thinout$mat)

remove.idx = which(colSums(Y)==0)

Y = log(Y[,-remove.idx]+eps)

X = model.matrix(~thinout$designmat)



mean(abs(thinout$coef) &lt; 10^-6)</code></pre>
<pre><code>[1] 0.899984</code></pre>
<pre class="r"><code>mout = mouthwash(Y,X,k=num_sv,cov_of_interest = 2,include_intercept = FALSE)</code></pre>
<pre><code>Running mouthwash on 709 x 2 matrix X and 709 x 12431 matrix Y.
 - Computing independent basis using QR decomposition.
 - Computation took 22.332 seconds.
 - Running additional preprocessing steps.
 - Computation took 0.001 seconds.
 - Running second step of mouthwash:
    + Estimating model parameters using EM.
    + Computation took 178.899 seconds.
    + Generating adaptive shrinkage (ash) output.
    + Computation took 1.437 seconds.
 - Second step took 181.393 seconds.
 - Estimating additional hidden confounders.
 - Computation took 2.378 seconds.</code></pre>
<pre class="r"><code>save(mout,file = &#39;data/mout_high.RData&#39;)
mout$pi0</code></pre>
<pre><code>[1] 0.9662333</code></pre>
<pre class="r"><code>#bout &lt;- backwash(Y = Y, X = X, k = num_sv, cov_of_interest = 2, include_intercept = FALSE)
#save(bout,file = &#39;data/bout_high.RData&#39;)
#bout$pi0


cate_cate = cate::cate.fit(X.primary = X[, 2, drop = FALSE], X.nuis = X[, -2, drop = FALSE],
                              Y = Y, r = num_sv, adj.method = &quot;rr&quot;)
save(cate_cate,file = &#39;data/cate_cate_high.RData&#39;)
sva_sva     &lt;- sva::sva(dat = t(Y), mod = X, mod0 = X[, -2, drop = FALSE], n.sv = num_sv)</code></pre>
<pre><code>Number of significant surrogate variables is:  3 
Iteration (out of 5 ):1  2  3  4  5  </code></pre>
<pre class="r"><code>save(sva_sva,file = &#39;data/sva_sva_high.RData&#39;)
X.sva &lt;- cbind(X, sva_sva$sv)
lmout &lt;- limma::lmFit(object = t(Y), design = X.sva)
eout  &lt;- limma::eBayes(lmout)
svaout           &lt;- list()
svaout$betahat   &lt;- lmout$coefficients[, 2]
svaout$sebetahat &lt;- lmout$stdev.unscaled[, 2] * sqrt(eout$s2.post)
svaout$pvalues   &lt;- eout$p.value[, 2]

which_null = c(1*(abs(thinout$coef[-remove.idx]) &lt; 10^-6))

roc_out &lt;- list(
  pROC::roc(response = which_null, predictor = c(mout$result$lfdr)),
  #pROC::roc(response = which_null, predictor = c(bout$result$lfdr)),
  pROC::roc(response = which_null, predictor = c(cate_cate$beta.p.value)),
  pROC::roc(response = which_null, predictor = c(svaout$pvalues)))
name_vec &lt;- c(&quot;MOUTHWASH&quot;, &quot;CATErr&quot;, &quot;SVA&quot;)
names(roc_out) &lt;- name_vec

sout &lt;- lapply(roc_out, function(x) { data.frame(TPR = x$sensitivities, FPR = 1 - x$specificities)})
for (index in 1:length(sout)) {
  sout[[index]]$Method &lt;- name_vec[index]
}
longdat &lt;- do.call(rbind, sout)

shortdat &lt;- dplyr::filter(longdat, Method == &quot;MOUTHWASH&quot; | 
                            Method == &quot;CATErr&quot; | Method == &quot;SVA&quot; | Method == &quot;LEAPP&quot;)
ggplot(data = shortdat, mapping = aes(x = FPR, y = TPR, col = Method)) +
  geom_path() + theme_bw() + ggtitle(&quot;ROC Curves&quot;)</code></pre>
<p><img src="figure/scdePBMC2.Rmd/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>auc_vec &lt;- sapply(roc_out, FUN = function(x) { x$auc })
knitr::kable(sort(auc_vec, decreasing = TRUE), col.names = &quot;AUC&quot;, digits = 3)</code></pre>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">AUC</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>SVA</td>
<td align="right">0.827</td>
</tr>
<tr class="even">
<td>CATErr</td>
<td align="right">0.814</td>
</tr>
<tr class="odd">
<td>MOUTHWASH</td>
<td align="right">0.685</td>
</tr>
</tbody>
</table>
<pre class="r"><code>method_list &lt;- list()
method_list$CATErr           &lt;- list()
method_list$CATErr$betahat   &lt;- c(cate_cate$beta)
method_list$CATErr$sebetahat &lt;- c(sqrt(cate_cate$beta.cov.row * c(cate_cate$beta.cov.col)) / sqrt(nrow(X)))

method_list$SVA             &lt;- list()
method_list$SVA$betahat     &lt;- c(svaout$betahat)
method_list$SVA$sebetahat   &lt;- c(svaout$sebetahat)

ashfit &lt;- lapply(method_list, FUN = function(x) { ashr::ash(x$betahat, x$sebetahat)})
api0 &lt;- sapply(ashfit, FUN = ashr::get_pi0)
api0 &lt;- c(api0, MOUTHWASH = mout$pi0)
#api0 &lt;- c(api0, BACKWASH = bout$pi0)

knitr::kable(sort(api0, decreasing = TRUE), col.names = &quot;Estimate of Pi0&quot;)</code></pre>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">Estimate of Pi0</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>MOUTHWASH</td>
<td align="right">0.9662333</td>
</tr>
<tr class="even">
<td>SVA</td>
<td align="right">0.4528195</td>
</tr>
<tr class="odd">
<td>CATErr</td>
<td align="right">0.1373782</td>
</tr>
</tbody>
</table>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.1 (2019-07-05)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] ggplot2_3.2.1               seqgendiff_1.1.1           
 [3] cate_1.1                    vicar_0.1-10               
 [5] ROTS_1.12.0                 Seurat_3.1.2               
 [7] MAST_1.10.0                 SingleCellExperiment_1.6.0 
 [9] SummarizedExperiment_1.14.1 DelayedArray_0.10.0        
[11] BiocParallel_1.18.1         matrixStats_0.55.0         
[13] Biobase_2.44.0              GenomicRanges_1.36.1       
[15] GenomeInfoDb_1.20.0         IRanges_2.18.3             
[17] S4Vectors_0.22.1            BiocGenerics_0.30.0        

loaded via a namespace (and not attached):
  [1] reticulate_1.13        R.utils_2.9.0          tidyselect_0.2.5      
  [4] lme4_1.1-21            RSQLite_2.1.2          AnnotationDbi_1.46.1  
  [7] htmlwidgets_1.5.1      grid_3.6.1             Rtsne_0.15            
 [10] pROC_1.15.3            munsell_0.5.0          codetools_0.2-16      
 [13] mutoss_0.1-12          ica_1.0-2              future_1.15.1         
 [16] withr_2.1.2            colorspace_1.4-1       leapp_1.2             
 [19] highr_0.8              knitr_1.25             pscl_1.5.2            
 [22] ROCR_1.0-7             gbRd_0.4-11            listenv_0.8.0         
 [25] labeling_0.3           Rdpack_0.11-0          git2r_0.26.1          
 [28] GenomeInfoDbData_1.2.1 mixsqp_0.1-97          mnormt_1.5-5          
 [31] bit64_0.9-7            rprojroot_1.3-2        vctrs_0.2.0           
 [34] TH.data_1.0-10         xfun_0.10              R6_2.4.0              
 [37] doParallel_1.0.15      rsvd_1.0.2             bitops_1.0-6          
 [40] assertthat_0.2.1       promises_1.1.0         SDMTools_1.1-221.2    
 [43] scales_1.0.0           multcomp_1.4-12        gtable_0.3.0          
 [46] npsurv_0.4-0           globals_0.12.5         sva_3.32.1            
 [49] sandwich_2.5-1         svd_0.5                workflowr_1.5.0       
 [52] rlang_0.4.0            zeallot_0.1.0          genefilter_1.66.0     
 [55] splines_3.6.1          lazyeval_0.2.2         yaml_2.2.0            
 [58] reshape2_1.4.3         abind_1.4-5            backports_1.1.5       
 [61] httpuv_1.5.2           tools_3.6.1            gplots_3.0.1.1        
 [64] RColorBrewer_1.1-2     ggridges_0.5.2         TFisher_0.2.0         
 [67] Rcpp_1.0.2             plyr_1.8.4             zlibbioc_1.30.0       
 [70] purrr_0.3.2            RCurl_1.95-4.12        pbapply_1.4-2         
 [73] ashr_2.2-38            cowplot_1.0.0          zoo_1.8-6             
 [76] ggrepel_0.8.1          cluster_2.1.0          fs_1.3.1              
 [79] magrittr_1.5           data.table_1.12.6      lmtest_0.9-37         
 [82] RANN_2.6.1             truncnorm_1.0-8        mvtnorm_1.0-11        
 [85] SQUAREM_2017.10-1      whisker_0.4            fitdistrplus_1.0-14   
 [88] lsei_1.2-0             evaluate_0.14          xtable_1.8-4          
 [91] XML_3.98-1.20          gridExtra_2.3          compiler_3.6.1        
 [94] tibble_2.1.3           KernSmooth_2.23-15     crayon_1.3.4          
 [97] minqa_1.2.4            R.oo_1.23.0            htmltools_0.4.0       
[100] mgcv_1.8-29            corpcor_1.6.9          later_1.0.0           
[103] tidyr_1.0.0            RcppParallel_4.4.4     DBI_1.0.0             
[106] MASS_7.3-51.4          boot_1.3-23            Matrix_1.2-17         
[109] R.methodsS3_1.7.1      gdata_2.18.0           metap_1.2             
[112] igraph_1.2.4.1         pkgconfig_2.0.3        sn_1.5-4              
[115] numDeriv_2016.8-1.1    plotly_4.9.1           foreach_1.4.7         
[118] annotate_1.62.0        blme_1.0-4             multtest_2.40.0       
[121] XVector_0.24.0         ruv_0.9.7.1            bibtex_0.4.2          
[124] stringr_1.4.0          digest_0.6.21          sctransform_0.2.1     
[127] RcppAnnoy_0.0.13       tsne_0.1-3             rmarkdown_1.16        
[130] leiden_0.3.1           uwot_0.1.5             gtools_3.8.1          
[133] nloptr_1.2.1           lifecycle_0.1.0        nlme_3.1-141          
[136] jsonlite_1.6           viridisLite_0.3.0      limma_3.40.6          
[139] pillar_1.4.2           lattice_0.20-38        httr_1.4.1            
[142] plotrix_3.7-7          survival_2.44-1.1      glue_1.3.1            
[145] esaBcv_1.2.1           png_0.1-7              iterators_1.0.12      
[148] bit_1.1-14             stringi_1.4.3          blob_1.2.0            
[151] memoise_1.1.0          caTools_1.17.1.2       dplyr_0.8.3           
[154] irlba_2.3.3            future.apply_1.4.0     ape_5.3               </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
