<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="DongyueXie" />

<meta name="date" content="2020-03-19" />

<title>scde Check Alpha 0/1 on real data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.0.13/css/fa-svg-with-js.css" rel="stylesheet" />
<script src="site_libs/font-awesome-5.0.13/js/fontawesome-all.min.js"></script>
<script src="site_libs/font-awesome-5.0.13/js/fa-v4-shims.min.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">misc</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/DongyueXie/misc">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">scde Check Alpha 0/1 on real data</h1>
<h4 class="author"><em>DongyueXie</em></h4>
<h4 class="date"><em>2020-03-19</em></h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-03-31
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>misc/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20191122code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20191122)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20191122code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20191122)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomDongyueXiemisctree4f066db82952abe4f1501cf0a860584285715bc7targetblank4f066dba"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/DongyueXie/misc/tree/4f066db82952abe4f1501cf0a860584285715bc7" target="_blank">4f066db</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomDongyueXiemisctree4f066db82952abe4f1501cf0a860584285715bc7targetblank4f066dba" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  analysis/contrainedclustering.Rmd
    Untracked:  analysis/gsea.Rmd
    Untracked:  analysis/ideas.Rmd
    Untracked:  analysis/methylation.Rmd
    Untracked:  analysis/susie.Rmd
    Untracked:  code/sccytokines.R
    Untracked:  code/scdeCalibration.R
    Untracked:  data/bart/
    Untracked:  data/cytokine/DE_controls_output_filter10.RData
    Untracked:  data/cytokine/DE_controls_output_filter10_addlimma.RData
    Untracked:  data/cytokine/README
    Untracked:  data/cytokine/test.RData
    Untracked:  data/cytokine_normalized.RData
    Untracked:  data/deconv/
    Untracked:  data/scde/

Unstaged changes:
    Modified:   analysis/binomthinultimate.Rmd
    Modified:   analysis/deconvolution.Rmd
    Modified:   analysis/index.Rmd
    Modified:   analysis/limma.Rmd
    Deleted:    data/mout_high.RData
    Deleted:    data/scCDT.RData
    Deleted:    data/sva_sva_high.RData

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DongyueXie/misc/blob/4f066db82952abe4f1501cf0a860584285715bc7/analysis/scdeCheckAlpha.Rmd" target="_blank">4f066db</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-03-31
</td>
<td>
wflow_publish(“analysis/scdeCheckAlpha.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/DongyueXie/misc/c0fdf5cfd82e0b951c63eecf89604a4d236b0108/docs/scdeCheckAlpha.html" target="_blank">c0fdf5c</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-03-19
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DongyueXie/misc/blob/a7a0098c82a6648ff8e311397c470d16a43dc588/analysis/scdeCheckAlpha.Rmd" target="_blank">a7a0098</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-03-19
</td>
<td>
wflow_publish(“analysis/scdeCheckAlpha.Rmd”)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>Compare the loglikelihood from sva-limma-ash, setting <span class="math inline">\(\alpha=0\)</span> and <span class="math inline">\(\alpha=1\)</span>, on real data.</p>
<div id="dataset-1-pbmc" class="section level2">
<h2>Dataset 1: PBMC</h2>
<div id="summary" class="section level3">
<h3>Summary</h3>
<p>Even if we do not take log, <span class="math inline">\(\alpha=1\)</span> still give larger loglikelihood.</p>
<pre class="r"><code>library(vicar)
library(sva)
library(cate)
library(seqgendiff)
library(edgeR)
library(MASS)
load(&#39;data/scde/scCD4.RData&#39;)
load(&#39;data/scde/scCD8.RData&#39;)
load(&#39;data/scde/scCD14.RData&#39;)
load(&#39;data/scde/scMB.RData&#39;)

mglmfit = function(Y,X){
  G = nrow(Y)
  coefs = c()
  ses = c()
  for(i in 1:G){
    fit = glm(y~.,data.frame(y=Y[i,],x=X[,-1]),family = &#39;poisson&#39;)
    coefs = rbind(coefs,fit$coefficients)
    ses = rbind(ses,summary(fit)$coefficients[,2])
  }
  return(list(coefficients=coefs,se = ses))
}</code></pre>
<p>CD4, CD8, CD14 and B cells. Look at number of genes and cells.</p>
<pre class="r"><code>dim(CD4)</code></pre>
<pre><code>[1] 13713   709</code></pre>
<pre class="r"><code>dim(CD8)</code></pre>
<pre><code>[1] 13713   313</code></pre>
<pre class="r"><code>dim(CD14)</code></pre>
<pre><code>[1] 13713   432</code></pre>
<pre class="r"><code>dim(MB)</code></pre>
<pre><code>[1] 13713   342</code></pre>
</div>
<div id="cd4-vs-mb-cells." class="section level3">
<h3>CD4 vs MB cells.</h3>
<pre class="r"><code>Y = as.matrix(cbind(CD4,MB))

group_idx = c(rep(1,dim(CD4)[2]),rep(0,dim(MB)[2]))

# remove genes appearing in less than 10 cells

Y = Y[-which(rowSums(Y!=0)&lt;10),]

X = model.matrix(~group_idx)

real_study = function(Y,X,run_sva_eb=TRUE,n.sv=NULL){
  
  if(run_sva_eb){
    sva_sva = sva((Y),mod=X,mod0=X[,1],n.sv = n.sv)
    X.sva &lt;- cbind(X, sva_sva$sv)
    lmout &lt;- limma::lmFit(object = (Y), design = X.sva)
    eout  &lt;- limma::eBayes(lmout)
    svaout           &lt;- list()
    svaout$betahat   &lt;- lmout$coefficients[, 2]
    svaout$sebetahat &lt;- lmout$stdev.unscaled[, 2] * sqrt(eout$s2.post)
    svaout$pvalues   &lt;- eout$p.value[, 2]
    sva_limma_ash0 = ashr::ash(svaout$betahat,svaout$sebetahat,alpha=0)
    sva_limma_ash1 = ashr::ash(svaout$betahat,svaout$sebetahat,alpha=1)
    print(paste(&#39;sva-limma-ash0 loglik: &#39;,round(sva_limma_ash0$loglik,2),&#39;;&#39;,
            &#39;sva-limma-ash1 loglik: &#39;,round(sva_limma_ash1$loglik)))
  }else{
    lmout &lt;- limma::lmFit(object = (Y), design = X)
    out = list()
    out$betahat   &lt;- lmout$coefficients[, 2]
    out$sebetahat &lt;- lmout$stdev.unscaled[, 2] * lmout$sigma
    lm_ash0 = ashr::ash(out$betahat,out$sebetahat,alpha=0)
    lm_ash1 = ashr::ash(out$betahat,out$sebetahat,alpha=1)
    print(paste(&#39;lm-ash0 loglik: &#39;,round(lm_ash0$loglik,2),&#39;;&#39;,
            &#39;lm-ash1 loglik: &#39;,round(lm_ash1$loglik)))
  }
}

real_study(Y,X,3)</code></pre>
<pre><code>Number of significant surrogate variables is:  3 
Iteration (out of 5 ):1  2  3  4  5  [1] &quot;sva-limma-ash0 loglik:  16134.95 ; sva-limma-ash1 loglik:  17053&quot;</code></pre>
<pre class="r"><code>real_study(Y,X,run_sva_eb = FALSE)</code></pre>
<pre><code>[1] &quot;lm-ash0 loglik:  16049.71 ; lm-ash1 loglik:  17018&quot;</code></pre>
<p>Take log1p of Y</p>
<pre class="r"><code>real_study(log(Y+0.5),X,14)</code></pre>
<pre><code>Number of significant surrogate variables is:  14 
Iteration (out of 5 ):1  2  3  4  5  [1] &quot;sva-limma-ash0 loglik:  16929.86 ; sva-limma-ash1 loglik:  17545&quot;</code></pre>
<pre class="r"><code>real_study(log(Y+0.5),X,run_sva_eb = FALSE)</code></pre>
<pre><code>[1] &quot;lm-ash0 loglik:  16628.98 ; lm-ash1 loglik:  17339&quot;</code></pre>
<p>Voom-lm-ash</p>
<pre class="r"><code>d = DGEList(Y)
d = calcNormFactors(d)
y = voom(d, design=X, plot = FALSE)
real_study(y,X,run_sva_eb = FALSE)</code></pre>
<pre><code>[1] &quot;lm-ash0 loglik:  2165.18 ; lm-ash1 loglik:  1471&quot;</code></pre>
<p>glm-ash</p>
<pre class="r"><code>fit = mglmfit(Y,X)
glm_ash0 = ashr::ash(fit$coefficients[,2],fit$se[,2],alpha=0)
glm_ash1 = ashr::ash(fit$coefficients[,2],fit$se[,2],alpha=1)
print(paste(&#39;glm-ash0 loglik: &#39;,round(glm_ash0$loglik,2),&#39;;&#39;,
            &#39;glm-ash1 loglik: &#39;,round(glm_ash1$loglik)))</code></pre>
<pre><code>[1] &quot;glm-ash0 loglik:  -9342.16 ; glm-ash1 loglik:  -9895&quot;</code></pre>
</div>
<div id="cd8-vs-cd14-cells" class="section level3">
<h3>CD8 vs CD14 cells</h3>
<pre class="r"><code>Y = as.matrix(cbind(CD8,CD14))

group_idx = c(rep(1,dim(CD8)[2]),rep(0,dim(CD14)[2]))

# remove genes appearing in less than 10 cells

Y = Y[-which(rowSums(Y!=0)&lt;10),]

X = model.matrix(~group_idx)
real_study(Y,X,3)</code></pre>
<pre><code>Number of significant surrogate variables is:  3 
Iteration (out of 5 ):1  2  3  4  5  [1] &quot;sva-limma-ash0 loglik:  14146.63 ; sva-limma-ash1 loglik:  15014&quot;</code></pre>
<pre class="r"><code>real_study(Y,X,run_sva_eb = FALSE)</code></pre>
<pre><code>[1] &quot;lm-ash0 loglik:  13372.89 ; lm-ash1 loglik:  14499&quot;</code></pre>
<p>Take log1p of Y</p>
<pre class="r"><code>real_study(log(Y+0.5),X,27)</code></pre>
<pre><code>Number of significant surrogate variables is:  27 
Iteration (out of 5 ):1  2  3  4  5  [1] &quot;sva-limma-ash0 loglik:  16140.94 ; sva-limma-ash1 loglik:  16406&quot;</code></pre>
<pre class="r"><code>real_study(log(Y+0.5),X,run_sva_eb = FALSE)</code></pre>
<pre><code>[1] &quot;lm-ash0 loglik:  14028.98 ; lm-ash1 loglik:  14877&quot;</code></pre>
<p>Voom-lm-ash</p>
<pre class="r"><code>d = DGEList(Y)
d = calcNormFactors(d)
y = voom(d, design=X, plot = FALSE)
real_study(y,X,run_sva_eb = FALSE)</code></pre>
<pre><code>[1] &quot;lm-ash0 loglik:  117.68 ; lm-ash1 loglik:  -1781&quot;</code></pre>
<p>glm-ash</p>
<pre class="r"><code>fit = mglmfit(Y,X)
glm_ash0 = ashr::ash(fit$coefficients[,2],fit$se[,2],alpha=0)
glm_ash1 = ashr::ash(fit$coefficients[,2],fit$se[,2],alpha=1)
print(paste(&#39;glm-ash0 loglik: &#39;,round(glm_ash0$loglik,2),&#39;;&#39;,
            &#39;glm-ash1 loglik: &#39;,round(glm_ash1$loglik)))</code></pre>
<pre><code>[1] &quot;glm-ash0 loglik:  -7016.82 ; glm-ash1 loglik:  -8182&quot;</code></pre>
</div>
</div>
<div id="dataset-2-jaitin-et-al.-2014" class="section level2">
<h2>Dataset 2: Jaitin et al. 2014</h2>
<p>Taken from <a href="https://github.com/jhsiao999/singleCellRNASeqMouseJaitinSpleen">here</a></p>
<pre class="r"><code>load(&#39;data/scde/MouseJaitinSpleen.rda&#39;)
suppressPackageStartupMessages(library(SummarizedExperiment))
suppressPackageStartupMessages(library(MultiAssayExperiment))
mat = exprs(MouseJaitinSpleen)
idx = which(MouseJaitinSpleen$ERCC_dilution==&#39;2.50E-05&#39;)
cell_type = MouseJaitinSpleen$group_name[idx]
table(cell_type)</code></pre>
<pre><code>cell_type
                B cell                 CD11c+        CD11c+(2hr_LPS) 
                    48                   2112                   1536 
         CD8-CD4+ESAM+                CD8-pDC              CD8+CD86- 
                    96                     96                     96 
               CD8+pDC              GC B cell monocyte_or_neutrophil 
                    96                     48                     48 
               NK_cell             splenocyte 
                    48                     48 </code></pre>
<pre class="r"><code>mat = mat[,idx]</code></pre>
<div id="b-cell-and-nk_cell" class="section level3">
<h3>B cell and NK_cell</h3>
<pre class="r"><code>Y = as.matrix(cbind(mat[,which(cell_type==&#39;B cell&#39;)],mat[,which(cell_type==&#39;NK_cell&#39;)]))

group_idx = c(rep(1,dim(mat[,which(cell_type==&#39;B cell&#39;)])[2]),
              rep(0,dim(mat[,which(cell_type==&#39;NK_cell&#39;)])[2]))

# remove genes appearing in less than 10 cells

Y = Y[-which(rowSums(Y!=0)&lt;10),]
rm.idx = which(colSums(Y)==0)
Y = Y[,-rm.idx]
group_idx = group_idx[-rm.idx]


X = model.matrix(~group_idx)
real_study(Y,X)</code></pre>
<pre><code>Number of significant surrogate variables is:  2 
Iteration (out of 5 ):1  2  3  4  5  [1] &quot;sva-limma-ash0 loglik:  -372.28 ; sva-limma-ash1 loglik:  -106&quot;</code></pre>
<pre class="r"><code>real_study(Y,X,run_sva_eb = FALSE)</code></pre>
<pre><code>[1] &quot;lm-ash0 loglik:  -388.83 ; lm-ash1 loglik:  -120&quot;</code></pre>
<pre class="r"><code># How about just run linear model and then ash?

X = model.matrix(~group_idx)

lmout &lt;- limma::lmFit(object = (Y), design = X)

svaout           &lt;- list()
svaout$betahat   &lt;- lmout$coefficients[, 2]
svaout$sebetahat &lt;- lmout$stdev.unscaled[, 2] * lmout$sigma
svaout$pvalues   &lt;- eout$p.value[, 2]

# sva-limma-ash-alpha 0

sva_limma_ash0 = ashr::ash(svaout$betahat,svaout$sebetahat,alpha=0)
sva_limma_ash0$loglik

# sva-limma-ash-alpha 1

sva_limma_ash1 = ashr::ash(svaout$betahat,svaout$sebetahat,alpha=1)
sva_limma_ash1$loglik</code></pre>
<p>Take log1p of y</p>
<pre class="r"><code>real_study(log(Y+0.5),X)</code></pre>
<pre><code>Number of significant surrogate variables is:  15 
Iteration (out of 5 ):1  2  3  4  5  [1] &quot;sva-limma-ash0 loglik:  304.17 ; sva-limma-ash1 loglik:  372&quot;</code></pre>
<pre class="r"><code>real_study(log(Y+0.5),X,run_sva_eb = FALSE)</code></pre>
<pre><code>[1] &quot;lm-ash0 loglik:  -147.77 ; lm-ash1 loglik:  -24&quot;</code></pre>
<p>Voom-lm-ash</p>
<pre class="r"><code>d = DGEList(Y)
d = calcNormFactors(d)
y = voom(d, design=X, plot = FALSE)
real_study(y,X,run_sva_eb = FALSE)</code></pre>
<pre><code>[1] &quot;lm-ash0 loglik:  -216.76 ; lm-ash1 loglik:  -173&quot;</code></pre>
<p>glm-ash</p>
<pre class="r"><code>fit = mglmfit(Y,X)
glm_ash0 = ashr::ash(fit$coefficients[,2],fit$se[,2],alpha=0)
glm_ash1 = ashr::ash(fit$coefficients[,2],fit$se[,2],alpha=1)
print(paste(&#39;glm-ash0 loglik: &#39;,round(glm_ash0$loglik,2),&#39;;&#39;,
            &#39;glm-ash1 loglik: &#39;,round(glm_ash1$loglik)))</code></pre>
<pre><code>[1] &quot;glm-ash0 loglik:  -2043.3 ; glm-ash1 loglik:  -2080&quot;</code></pre>
</div>
<div id="cd8-pdc-and-splenocyte" class="section level3">
<h3>CD8-pDC and splenocyte</h3>
<pre class="r"><code>Y = as.matrix(cbind(mat[,which(cell_type==&#39;CD8-pDC&#39;)],mat[,which(cell_type==&#39;splenocyte&#39;)]))

group_idx = c(rep(1,dim(mat[,which(cell_type==&#39;CD8-pDC&#39;)])[2]),
              rep(0,dim(mat[,which(cell_type==&#39;splenocyte&#39;)])[2]))

# remove genes appearing in less than 10 cells

Y = Y[-which(rowSums(Y!=0)&lt;10),]
rm.idx = which(colSums(Y)==0)
Y = Y[,-rm.idx]
group_idx = group_idx[-rm.idx]

X = model.matrix(~group_idx)
real_study(Y,X)</code></pre>
<pre><code>Number of significant surrogate variables is:  3 
Iteration (out of 5 ):1  2  3  4  5  [1] &quot;sva-limma-ash0 loglik:  186.59 ; sva-limma-ash1 loglik:  539&quot;</code></pre>
<pre class="r"><code>real_study(Y,X,run_sva_eb = FALSE)</code></pre>
<pre><code>[1] &quot;lm-ash0 loglik:  97.49 ; lm-ash1 loglik:  488&quot;</code></pre>
<p>Take log1p of y</p>
<pre class="r"><code>real_study(log(Y+0.5),X)</code></pre>
<pre><code>Number of significant surrogate variables is:  24 
Iteration (out of 5 ):1  2  3  4  5  [1] &quot;sva-limma-ash0 loglik:  927.11 ; sva-limma-ash1 loglik:  1011&quot;</code></pre>
<pre class="r"><code>real_study(log(Y+0.5),X,run_sva_eb = FALSE)</code></pre>
<pre><code>[1] &quot;lm-ash0 loglik:  301.85 ; lm-ash1 loglik:  519&quot;</code></pre>
<p>Voom-lm-ash</p>
<pre class="r"><code>d = DGEList(Y)
d = calcNormFactors(d)
y = voom(d, design=X, plot = FALSE)
real_study(y,X,run_sva_eb = FALSE)</code></pre>
<pre><code>[1] &quot;lm-ash0 loglik:  -122.92 ; lm-ash1 loglik:  -12&quot;</code></pre>
<p>glm-ash</p>
<pre class="r"><code>fit = mglmfit(Y,X)
glm_ash0 = ashr::ash(fit$coefficients[,2],fit$se[,2],alpha=0)
glm_ash1 = ashr::ash(fit$coefficients[,2],fit$se[,2],alpha=1)
print(paste(&#39;glm-ash0 loglik: &#39;,round(glm_ash0$loglik,2),&#39;;&#39;,
            &#39;glm-ash1 loglik: &#39;,round(glm_ash1$loglik)))</code></pre>
<pre><code>[1] &quot;glm-ash0 loglik:  -3605.35 ; glm-ash1 loglik:  -3773&quot;</code></pre>
</div>
</div>
<div id="dataset3-grun-et-al-2016" class="section level2">
<h2>Dataset3: Grun et al 2016</h2>
<p>Taken <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE81076">here</a></p>
<pre class="r"><code>datax=readRDS(&quot;data/scde/GSE81076-GPL18573.rds&quot;)
datax_gene = experiments(datax)[[&quot;gene&quot;]]
mat = (assays(datax_gene)[[&quot;count&quot;]])
cell63_idx = 1:96
TGFBR3_idx = 673:768
Y = cbind(mat[,cell63_idx],mat[,TGFBR3_idx])
group_idx = c(rep(1,length(cell63_idx)),
              rep(0,length(TGFBR3_idx)))
Y = Y[-which(rowSums(Y!=0)&lt;10),]

X = model.matrix(~group_idx)
real_study(Y,X)</code></pre>
<pre><code>Number of significant surrogate variables is:  2 
Iteration (out of 5 ):1  2  3  4  5  [1] &quot;sva-limma-ash0 loglik:  -5114.92 ; sva-limma-ash1 loglik:  -2264&quot;</code></pre>
<pre class="r"><code>real_study(Y,X,run_sva_eb = FALSE)</code></pre>
<pre><code>[1] &quot;lm-ash0 loglik:  -5100.36 ; lm-ash1 loglik:  -2259&quot;</code></pre>
<p>Take log1p of y</p>
<pre class="r"><code>real_study(log(Y+0.5),X)</code></pre>
<pre><code>Number of significant surrogate variables is:  30 
Iteration (out of 5 ):1  2  3  4  5  [1] &quot;sva-limma-ash0 loglik:  2766.62 ; sva-limma-ash1 loglik:  3803&quot;</code></pre>
<pre class="r"><code>real_study(log(Y+0.5),X,run_sva_eb = FALSE)</code></pre>
<pre><code>[1] &quot;lm-ash0 loglik:  160.62 ; lm-ash1 loglik:  1706&quot;</code></pre>
<p>Voom-lm-ash</p>
<pre class="r"><code>d = DGEList(Y)
d = calcNormFactors(d)
y = voom(d, design=X, plot = FALSE)
real_study(y,X,run_sva_eb = FALSE)</code></pre>
<pre><code>[1] &quot;lm-ash0 loglik:  -7375.07 ; lm-ash1 loglik:  -7302&quot;</code></pre>
<p>glm-ash</p>
<pre class="r"><code>fit = mglmfit(Y,X)
glm_ash0 = ashr::ash(fit$coefficients[,2],fit$se[,2],alpha=0)
glm_ash1 = ashr::ash(fit$coefficients[,2],fit$se[,2],alpha=1)
print(paste(&#39;glm-ash0 loglik: &#39;,round(glm_ash0$loglik,2),&#39;;&#39;,
            &#39;glm-ash1 loglik: &#39;,round(glm_ash1$loglik)))</code></pre>
<pre><code>[1] &quot;glm-ash0 loglik:  -15292.75 ; glm-ash1 loglik:  -16158&quot;</code></pre>
</div>
<div id="also-compare-log-limma-ash-voom-limma-ash-and-glm-ash-on-simulated-and-real-data." class="section level2">
<h2>Also compare log-limma-ash, voom-limma-ash and glm-ash on simulated and real data.</h2>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.5.1 (2018-07-02)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Scientific Linux 7.4 (Nitrogen)

Matrix products: default
BLAS/LAPACK: /software/openblas-0.2.19-el7-x86_64/lib/libopenblas_haswellp-r0.2.19.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] MultiAssayExperiment_1.8.3  SummarizedExperiment_1.12.0
 [3] DelayedArray_0.8.0          matrixStats_0.54.0         
 [5] Biobase_2.42.0              GenomicRanges_1.34.0       
 [7] GenomeInfoDb_1.18.1         IRanges_2.16.0             
 [9] S4Vectors_0.20.1            BiocGenerics_0.28.0        
[11] MASS_7.3-51.1               edgeR_3.24.0               
[13] limma_3.38.2                seqgendiff_1.2.1           
[15] cate_1.0.4                  sva_3.30.0                 
[17] BiocParallel_1.16.0         genefilter_1.64.0          
[19] mgcv_1.8-25                 nlme_3.1-137               
[21] vicar_0.1-10               

loaded via a namespace (and not attached):
 [1] svd_0.4.1              bitops_1.0-6           fs_1.3.1              
 [4] bit64_0.9-7            doParallel_1.0.14      rprojroot_1.3-2       
 [7] tools_3.5.1            backports_1.1.2        R6_2.3.0              
[10] DBI_1.0.0              lazyeval_0.2.1         colorspace_1.3-2      
[13] tidyselect_0.2.5       gridExtra_2.3          bit_1.1-14            
[16] compiler_3.5.1         git2r_0.26.1           scales_1.0.0          
[19] SQUAREM_2017.10-1      mixsqp_0.2-2           stringr_1.3.1         
[22] esaBcv_1.2.1           digest_0.6.18          rmarkdown_1.10        
[25] XVector_0.22.0         pscl_1.5.2             pkgconfig_2.0.2       
[28] htmltools_0.3.6        ruv_0.9.7              rlang_0.4.0           
[31] RSQLite_2.1.1          dplyr_0.8.0.1          leapp_1.2             
[34] RCurl_1.95-4.11        magrittr_1.5           GenomeInfoDbData_1.2.0
[37] Matrix_1.2-15          Rcpp_1.0.2             munsell_0.5.0         
[40] stringi_1.2.4          whisker_0.3-2          yaml_2.2.0            
[43] zlibbioc_1.28.0        plyr_1.8.4             grid_3.5.1            
[46] blob_1.1.1             promises_1.0.1         crayon_1.3.4          
[49] lattice_0.20-38        splines_3.5.1          annotate_1.60.0       
[52] locfit_1.5-9.1         knitr_1.20             pillar_1.3.1          
[55] corpcor_1.6.9          codetools_0.2-15       XML_3.98-1.16         
[58] glue_1.3.0             evaluate_0.12          httpuv_1.4.5          
[61] foreach_1.4.4          gtable_0.2.0           purrr_0.3.2           
[64] assertthat_0.2.0       ashr_2.2-39            ggplot2_3.1.1         
[67] xtable_1.8-3           later_0.7.5            survival_2.43-1       
[70] truncnorm_1.0-8        tibble_2.1.1           iterators_1.0.10      
[73] AnnotationDbi_1.44.0   memoise_1.1.0          workflowr_1.6.0       </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
