<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="DongyueXie" />

<meta name="date" content="2020-09-22" />

<title>SymSim</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.0.13/css/fa-svg-with-js.css" rel="stylesheet" />
<script src="site_libs/font-awesome-5.0.13/js/fontawesome-all.min.js"></script>
<script src="site_libs/font-awesome-5.0.13/js/fa-v4-shims.min.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">misc</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/DongyueXie/misc">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">SymSim</h1>
<h4 class="author"><em>DongyueXie</em></h4>
<h4 class="date"><em>2020-09-22</em></h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-09-23
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>misc/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20191122code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20191122)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20191122code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20191122)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomDongyueXiemisctreed2b54b194819994c0a24c165a3a049344e090c61targetblankd2b54b1a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/DongyueXie/misc/tree/d2b54b194819994c0a24c165a3a049344e090c61" target="_blank">d2b54b1</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomDongyueXiemisctreed2b54b194819994c0a24c165a3a049344e090c61targetblankd2b54b1a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/DongyueXie/misc/tree/d2b54b194819994c0a24c165a3a049344e090c61" target="_blank">d2b54b1</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  analysis/deconvSimulation2.Rmd
    Untracked:  analysis/ideas.Rmd
    Untracked:  code/sccytokines.R
    Untracked:  code/scdeCalibration.R
    Untracked:  data/bart/
    Untracked:  data/cytokine/DE_controls_output_filter10.RData
    Untracked:  data/cytokine/DE_controls_output_filter10_addlimma.RData
    Untracked:  data/cytokine/README
    Untracked:  data/cytokine/test.RData
    Untracked:  data/cytokine_normalized.RData
    Untracked:  data/deconv/
    Untracked:  data/scde/

Unstaged changes:
    Modified:   analysis/deconvSimulation.Rmd
    Modified:   analysis/deconvolution.Rmd
    Deleted:    data/mout_high.RData
    Deleted:    data/scCDT.RData
    Deleted:    data/sva_sva_high.RData

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/SymSim.Rmd</code>) and HTML (<code>docs/SymSim.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/DongyueXie/misc/blob/d2b54b194819994c0a24c165a3a049344e090c61/analysis/SymSim.Rmd" target="_blank">d2b54b1</a>
</td>
<td>
DongyueXie
</td>
<td>
2020-09-23
</td>
<td>
wflow_publish(c(“analysis/index.Rmd”, “analysis/genelength.Rmd”,
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="overview" class="section level2">
<h2>Overview</h2>
<p><a href="https://www.nature.com/articles/s41467-019-10500-w#Sec13">SymSim</a> is a method simulating single cell data. I find it comprehensive, generative and mimic the experimental procedure.</p>
<p><a href="https://github.com/YosefLab/SymSim">github</a>, <a href="SymSimTutorial.html">tutorial</a></p>
<p>Considers three variations: extrinsic variation(cell type, cell state), intrinsic variation(promoter on/off, mRNA synthesis rate, degradation), and technical variation(library preparation and sequencing).</p>
<p>Geneate counts: 1. generate transcript counts; 2. generate observed expressions.</p>
<p>The true transcript counts are generated using classical promoter kinetic model, with three parameters: promotoer on rate(<span class="math inline">\(k_{on}\)</span>), promoter off rate(<span class="math inline">\(k_{off}\)</span>) and RNA synthesis rate(<span class="math inline">\(s\)</span>). The value of these parameters are determined by the product of gene effects and cell specific extrinsic variability EVF(indicate the cell state, low dimensional manifold). EVF value is determined by a tree stucture(dictates cell-cell similarity map). For homogenous population(a single location in a tree, like one cell type at a certain state/time), EVFs are drawn iid from a distribution.</p>
</div>
<div id="generate-transcript-counts" class="section level2">
<h2>Generate transcript counts</h2>
<p>Two-state kinetic model: promoter switches between an on and an off states with certain probabilities. Let <span class="math inline">\(s\)</span> be the transcript rate and <span class="math inline">\(d\)</span> be the degradation rate fixed at 1. The stationary distribution for each gene analytically is a Beta-Poisson mixture.</p>
<p>Suppose there are <span class="math inline">\(m\)</span> genes and <span class="math inline">\(n\)</span> cells. Then there are three <span class="math inline">\(m\times n\)</span> matrices: <span class="math inline">\(K_{on}\)</span>, <span class="math inline">\(K_{off}, S\)</span>, each is a product of two low rank matrix. When the gene is on it is transcribed with transcription rate <span class="math inline">\(s\)</span>. Given the three matrices, transcript counts are drawn from a beta-Poisson model.</p>
<p>Draw transcript counts: for each gene in a cell, 1. generate <span class="math inline">\(p\)</span> from <span class="math inline">\(Beta(k_{on},k_{on})\)</span>; 2, generate transcript counts from <span class="math inline">\(Poisson(p*s)\)</span></p>
</div>
<div id="generate-observed-expressions" class="section level2">
<h2>Generate observed expressions</h2>
<p>cell efficiency, amplification bias, fragmentation, sequencing.</p>
<pre class="r"><code>library(SymSim)
cal_amp_bias &lt;- function(lenslope, nbins, gene_len, amp_bias_limit){
  
  ngenes &lt;- length(gene_len)
  len_bias_bin &lt;- (-c(1:nbins))*lenslope
  len_bias_bin &lt;- len_bias_bin-median(len_bias_bin)
  if (max(len_bias_bin) &gt; amp_bias_limit[2]) {
    stop(&quot;The lenslope parameter is too large.&quot;)
  }
  max_rand_bias &lt;- amp_bias_limit[2] - max(len_bias_bin)
  
  rand_bias &lt;- rnorm(ngenes, mean=0, sd=max_rand_bias)
  rand_bias[rand_bias &gt; max_rand_bias] &lt;- max_rand_bias
  rand_bias[rand_bias &lt; -max_rand_bias] &lt;- -max_rand_bias
  #rand_bias &lt;- runif(ngenes, -max_rand_bias,  max_rand_bias)
  
  binsize &lt;- floor(ngenes/nbins)
  genes_in_bins &lt;- vector(&quot;list&quot;, nbins)
  bin4genes &lt;- numeric(ngenes)
  for (ibin in 1:(nbins-1)){
    genes_in_bins[[ibin]] &lt;- order(gene_len)[((ibin-1)*binsize+1) : (ibin*binsize)]
    bin4genes[genes_in_bins[[ibin]]] &lt;- ibin
  }
  genes_in_bins[[nbins]] &lt;- order(gene_len)[((nbins-1)*binsize+1) : ngenes]
  bin4genes[genes_in_bins[[nbins]]] &lt;- nbins
  
  len_bias &lt;- numeric(ngenes); len_bias &lt;- len_bias_bin[bin4genes]
  amp_bias &lt;- rand_bias+len_bias
  return(amp_bias)
}

expand2binary &lt;- function(true_counts_1cell){
  expanded_vec &lt;- rep(1, sum(true_counts_1cell))
  trans_idx &lt;- sapply(which(true_counts_1cell&gt;0), 
                      function(igene){return(rep(igene, true_counts_1cell[igene]))})
  trans_idx &lt;- unlist(trans_idx)
  return(list(expanded_vec, trans_idx))
}


data(&quot;gene_len_pool&quot;)
true_counts_res &lt;- SimulateTrueCounts(ncells_total=100, ngenes=100, evf_type=&quot;one.population&quot;, Sigma=0.4, randseed=0)
true_counts_1cell = true_counts_res$counts[,1]
protocol = &#39;nonUMI&#39; 
rate_2cap= 0.1
gene_len = sample(gene_len_pool,100)
amp_bias = cal_amp_bias(0.02, 20, gene_len, c(-0.2, 0.2))
rate_2PCR=0.8 
nPCR1=16 
nPCR2=10 
LinearAmp=FALSE 
N_molecules_SEQ = 1e4

ngenes &lt;- length(gene_len)
if (protocol==&quot;nonUMI&quot;){data(len2nfrag)} 
inds &lt;- vector(&quot;list&quot;,2)
expanded_res &lt;- expand2binary(c(true_counts_1cell,1))
expanded_vec &lt;- expanded_res[[1]]; trans_idx &lt;- expanded_res[[2]]
inds[[1]] &lt;- which(expanded_vec &gt; 0); expanded_vec &lt;- expanded_vec[inds[[1]]]
trans_idx &lt;- trans_idx[inds[[1]]]

captured_vec &lt;- expanded_vec; captured_vec[runif(length(captured_vec)) &gt; rate_2cap] &lt;- 0
captured_vec[length(captured_vec)] &lt;- 1
inds[[2]] &lt;- which(captured_vec &gt; 0); captured_vec &lt;- captured_vec[inds[[2]]]
trans_idx &lt;- trans_idx[inds[[2]]]
amp_rate &lt;- c((rate_2PCR+amp_bias[trans_idx[1:(length(trans_idx)-1)]]),1)

## what does this step do?

temp &lt;- runif(length(captured_vec)) &lt; amp_rate
    temp &lt;- temp*2+captured_vec-temp
    for (iPCR in 2:nPCR1){
      eff &lt;- runif(length(temp))*amp_rate
      v1 &lt;- temp*(1-eff)
      round_down &lt;- (v1-floor(v1)) &lt; runif(length(v1))
      v1[round_down] &lt;- floor(v1[round_down]); v1[!round_down] &lt;- ceiling(v1[!round_down])
      temp &lt;- v1 + 2*(temp-v1)
    }
    PCRed_vec &lt;- temp</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.5.1 (2018-07-02)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Scientific Linux 7.4 (Nitrogen)

Matrix products: default
BLAS/LAPACK: /software/openblas-0.2.19-el7-x86_64/lib/libopenblas_haswellp-r0.2.19.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
 [1] stats4    parallel  grid      stats     graphics  grDevices utils    
 [8] datasets  methods   base     

other attached packages:
 [1] SymSim_0.0.0.9000           SummarizedExperiment_1.12.0
 [3] DelayedArray_0.8.0          BiocParallel_1.16.0        
 [5] matrixStats_0.54.0          GenomicRanges_1.34.0       
 [7] GenomeInfoDb_1.18.1         IRanges_2.16.0             
 [9] S4Vectors_0.20.1            repr_0.17                  
[11] phytools_0.6-99             maps_3.3.0                 
[13] roxygen2_6.1.1              stringi_1.2.4              
[15] MASS_7.3-51.1               ape_5.2                    
[17] Biobase_2.42.0              BiocGenerics_0.28.0        
[19] RColorBrewer_1.1-2          reshape_0.8.8              
[21] Rtsne_0.15                  ggplot2_3.1.1              
[23] plyr_1.8.4                 

loaded via a namespace (and not attached):
 [1] gtools_3.8.1            assertthat_0.2.0       
 [3] expm_0.999-3            animation_2.5          
 [5] GenomeInfoDbData_1.2.0  yaml_2.2.0             
 [7] numDeriv_2016.8-1       pillar_1.3.1           
 [9] backports_1.1.2         lattice_0.20-38        
[11] glue_1.3.0              quadprog_1.5-5         
[13] phangorn_2.5.5          digest_0.6.18          
[15] promises_1.0.1          XVector_0.22.0         
[17] colorspace_1.3-2        htmltools_0.3.6        
[19] httpuv_1.4.5            Matrix_1.2-15          
[21] pkgconfig_2.0.2         zlibbioc_1.28.0        
[23] purrr_0.3.2             scales_1.0.0           
[25] whisker_0.3-2           later_0.7.5            
[27] git2r_0.26.1            tibble_2.1.1           
[29] combinat_0.0-8          withr_2.1.2            
[31] lazyeval_0.2.1          mnormt_1.5-5           
[33] magrittr_1.5            crayon_1.3.4           
[35] evaluate_0.12           fs_1.3.1               
[37] nlme_3.1-137            xml2_1.2.0             
[39] tools_3.5.1             stringr_1.3.1          
[41] munsell_0.5.0           plotrix_3.7-4          
[43] compiler_3.5.1          clusterGeneration_1.3.4
[45] rlang_0.4.0             RCurl_1.95-4.11        
[47] igraph_1.2.2            bitops_1.0-6           
[49] base64enc_0.1-3         rmarkdown_1.10         
[51] gtable_0.2.0            R6_2.3.0               
[53] knitr_1.20              dplyr_0.8.0.1          
[55] fastmatch_1.1-0         commonmark_1.6         
[57] workflowr_1.6.2         rprojroot_1.3-2        
[59] Rcpp_1.0.4.6            scatterplot3d_0.3-41   
[61] tidyselect_0.2.5        coda_0.19-2            </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
