---
title: "voom and limma"
author: "Dongyue Xie"
date: "2020-01-04"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Introduction

[limma tutorial](https://ucdavis-bioinformatics-training.github.io/2018-June-RNA-Seq-Workshop/thursday/DE.html)

The number of reads observed for a given gene is proportional not just to the expression level of the gene but
also to its gene transcript length and to the sequencing
depth of the library. Dividing each read count by the corresponding library size (in millions) yields counts per million (cpm), a simple measure of read abundance that can
be compared across libraries of different sizes. Standardizing further by transcript length (in kilobases) gives rise
to reads per kilobase per million (rpkm).

Differences in log-cpm between samples can be
interpreted as log-fold-changes of expression. The counts
are augmented by a small positive value (a half of one
read) to avoid taking the logarithm of zero. This ensures
no missing log-cpm values and reduces the variability at
low count values.

Let $r_{gi}$ denote the read count of gene $g$ in sample $i$, and $y_{gi} = \log_2 (\frac{r_{gi}+0.5}{R_i+1}*10^6)$, where $R_i = \sum_gr_{gi}$. Voom assumes mean-variance relationship of $r$ being $var(r) = \lambda+\phi\lambda^2$, where $\lambda=E(r)$.  

```{r,eval=FALSE}
library(edgeR)
counts = read.delim('https://raw.githubusercontent.com/ucdavis-bioinformatics-training/2018-June-RNA-Seq-Workshop/master/thursday/all_counts.txt')
head(counts)

d0 <- DGEList(counts)

d0 <- calcNormFactors(d0)

cutoff <- 1
drop <- which(apply(cpm(d0), 1, max) < cutoff)
d <- d0[-drop,] 
dim(d) # number of genes left

snames <- colnames(counts) # Sample names

cultivar <- substr(snames, 1, nchar(snames) - 2) 
time <- substr(snames, nchar(snames) - 1, nchar(snames) - 1)

group <- interaction(cultivar, time)

mm <- model.matrix(~0 + group)

y <- voom(d, mm, plot = T)

# lmFit fits a linear model using weighted least squares for each gene:

fit <- lmFit(y, mm)

head(coef(fit))

dim(coef(fit))

contr <- makeContrasts(groupI5.9 - groupI5.6, levels = colnames(coef(fit)))
contr
```

