---
title: "limma"
author: "Dongyue Xie"
date: "2020-01-04"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Introduction

[limma tutorial](https://ucdavis-bioinformatics-training.github.io/2018-June-RNA-Seq-Workshop/thursday/DE.html)

```{r,eval=FALSE}
library(edgeR)
counts = read.delim('https://raw.githubusercontent.com/ucdavis-bioinformatics-training/2018-June-RNA-Seq-Workshop/master/thursday/all_counts.txt')
head(counts)

d0 <- DGEList(counts)

d0 <- calcNormFactors(d0)

cutoff <- 1
drop <- which(apply(cpm(d0), 1, max) < cutoff)
d <- d0[-drop,] 
dim(d) # number of genes left

snames <- colnames(counts) # Sample names

cultivar <- substr(snames, 1, nchar(snames) - 2) 
time <- substr(snames, nchar(snames) - 1, nchar(snames) - 1)

group <- interaction(cultivar, time)

mm <- model.matrix(~0 + group)

y <- voom(d, mm, plot = T)

# lmFit fits a linear model using weighted least squares for each gene:

fit <- lmFit(y, mm)

head(coef(fit))

dim(coef(fit))

contr <- makeContrasts(groupI5.9 - groupI5.6, levels = colnames(coef(fit)))
contr
```

