---
title: "Single cell cytokines data"
author: "DongyueXie"
date: "2020-02-17"
output:
  workflowr::wflow_html:
    code_folding: hide
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE,warning = FALSE)
```

## Introduction

See if sva works when comparing two control groups in single cell cytokine study.

Take a look at the annotation file:

```{r}
data_ann = read.csv('/project2/mstephens/chevrier-stephens/data/whole_cyto_annot.csv.gz')
data_ann[1:5,]
```

sample: treatment or cytokines, including two controls.

well?  There are 25 wells, each of which contains two samples.

cell_type: focus on 8 types of cells, B_cells, CD4_T_cells, CD8_T_cells, NK_cells, Dendritic-cells, Ly6C+, Ly6C-,Neutrolphils.

```{r}
#read data matrix
# library(hdf5r)
# library(Matrix)
# f <- H5File$new("whole_cyto_normalized.h5ad",mode = "r")
# print(names(f))
# out <- f[["X"]]
# print(h5attributes(out))
# i <- out[["indices"]][]
# j <- out[["indptr"]][]
# x <- out[["data"]][]
#library(SparseM)
# X.csr = new('matrix.csr',ra=x,ja=as.integer(i+1),ia=as.integer(j+1),dimension=h5attributes(out)$h5sparse_shape)
#load("data/cytokine_normalized.RData")
load('data/cytokine/DE_controls_output.RData')
```

Number of cells in each control group:

```{r}
n.sample = c()
for(cell in names(output)){
  n.sample = rbind(n.sample,c(sum(output[[cell]]$group_idx),sum(1-output[[cell]]$group_idx)))
}
rownames(n.sample) = names(output)
colnames(n.sample) = c('Ctrl-1','Ctrl-2')
knitr::kable(n.sample,caption = 'Number of samples')
```

Number of genes that have at least one measurement in two control groups.

```{r}
n.gene = c(14853)
for(cell in names(output)){
  n.gene = rbind(n.gene,c(14853-length(output[[cell]]$rm.idx)))
}
rownames(n.gene) = c('Total',names(output))
knitr::kable(n.gene,caption = 'Number of genes considered',col.names = '#genes')
```


## Correlations between PCs and groups

Plot of correlations between groups and first 20 principle components.

```{r,fig.width=9,fig.height=7}
par(mfrow=c(3,3))
for(cell in names(output)){
  plot(output[[cell]]$pc.cor[1:20],xlab='PCs',ylab='corr',main=paste(cell),ylim = c(-0.5,0.6),pch=20)
  abline(h=0,lty=3)
}
```


Plot of principle component that has maximum absolute correlation with groups for each cell. Vertical line seperates two groups.

```{r,fig.width=9,fig.height=7}
par(mfrow=c(3,3))
for(cell in names(output)){
  plot(output[[cell]]$pc.cor.max,xlab='',ylab='',
       main=paste(cell,', PC:',which.max(abs(output[[cell]]$pc.cor)),', corr:',round(max(abs(output[[cell]]$pc.cor)),2),sep=''),pch=1,col='grey50')
  abline(v=sum(output[[cell]]$group_idx),lty=3)
}
```

## Compare p-value distributions, t-test and sva-limma

Note: We will see a lot of p-values from t-test around 0.3-0.4. In single cell DE study, some genes are only measured once in one group while have no observation in another group. For example, gene expression in group 1 $= (0,0,...,0,0)$ and gene expression in group 2 $= (x,0,...,0,0)$. So in this case, unequal variance two-sample t-test always gives t-statistic = $1$ with $df=n_2-1$, where $n_2$ is the number of samples in group 2. Let's plot p-value vs df.

Suppose we have at least 5 samples in group 2, then p-value starts at 0.3739(df=4) and converges to 0.3173 as df goes to infinite.

```{r}
par(mfrow=c(1,1))
plot(4:1e3,(1-pt(1,4:1e3))*2,xlab='df',ylab='p-value',main='t-statistics = 1',pch=20)
```


Now compare distributions of p-values from t-test and sva-limma:

```{r,fig.width=6,fig.height=20}
par(mfrow=c(8,2))
for(cell in names(output)){
  hist(output[[cell]]$pvalue_t,main=paste(cell, ',t test'),xlab='')
  hist(output[[cell]]$pvalue_sva_limma,main=paste(cell, ',sva-limma'),xlab='')
}
```

## Compare the number of significant genes

The number of significant genes at $fdr=0.05$ by BH procedure.

```{r}
par(mfrow=c(1,1))
n.sig = c()
for(cell in names(output)){
  n.sig = rbind(n.sig,c(length(output[[cell]]$rej.idx.ttest),length(output[[cell]]$rej.idx.sva)))
}
rownames(n.sig) = names(output)
colnames(n.sig) = c('t-test','sva')
knitr::kable(n.sig,caption = 'Number of significant genes at fdr=0.05')
```


The number of significant genes at $fdr=0.01$ by BH procedure.

```{r}
BH = function(p,alpha=0.05){
  n=length(p)
  i=rank(p)
  idx = which(p<=(i/n*alpha))
  if(length(idx)==0){
    NULL
  }else{
    i0= max(i[idx])
    rej.idx = which(i<=i0)
    rej.idx
  }
}

n.sig = c()
for(cell in names(output)){
  n.sig = rbind(n.sig,c(length(BH(output[[cell]]$pvalue_t,0.01)),length(BH(output[[cell]]$pvalue_sva_limma,0.01))))
}
rownames(n.sig) = names(output)
colnames(n.sig) = c('t-test','sva')
knitr::kable(n.sig,caption = 'Number of significant genes at fdr=0.01')
```

